// @composio/core/dist/cancellation-W65QRyj2.mjs
import { u as logger } from "./SsrfErrors-BQRozkeU.mjs";
import { f as ComposioRequestCancelledError, g as JsonSchemaToZodError, h as JsonSchemaRefResolutionError, m as isRequestAbortError } from "./errors-B165yJUQ.mjs";
import { t as isPlainObject } from "./FileToolModifier.utils.neutral-B5rT711M.mjs";
import { jsonSchemaToZod } from "@composio/json-schema-to-zod";

//#region src/utils/jsonSchema.ts
const MAX_REF_CHAIN_DEPTH = 100;
const MAX_NODE_DEPTH = 512;
const CYCLE_BREAK_SENTINEL = {
	type: "object",
	additionalProperties: true
};
/**
* In-band hint attached to the cycle-break sentinel when lenient mode
* substitutes it for a dangling `$ref`. Makes the degradation visible to
* LLMs that read the wrapped tool's schema — without it the LLM sees a
* useless permissive object and no prose context. Wording is intentionally
* context-neutral ("Schema shape", not "Output shape") because the helper
* is used for both `inputParameters` and `outputParameters`. Overridden
* when the caller's `$ref` node carries its own `description` sibling
* (Draft 2020-12 sibling-keyword semantics).
*/
const UNRESOLVED_REF_DESCRIPTION = "Schema shape unresolved at the source — validate loosely. See https://github.com/ComposioHQ/composio/issues/3307.";
const POLLUTING_KEYS = /* @__PURE__ */ new Set([
	"__proto__",
	"constructor",
	"prototype"
]);
const REF_RESOLUTION_FIXES = ["Ensure the $ref pointer matches a path in $defs or definitions", "External $ref pointers (http://, https://, file://, …) are not resolved by the SDK"];
const decodePointerSegment = (segment) => segment.replace(/~1/g, "/").replace(/~0/g, "~");
const tryStep = (cursor, segment) => {
	if (cursor === null || typeof cursor !== "object") return {
		kind: "unresolved",
		reason: "missing-target",
		failedAt: segment
	};
	if (Array.isArray(cursor)) {
		const i = Number(segment);
		if (!Number.isInteger(i) || i < 0 || i >= cursor.length) return {
			kind: "unresolved",
			reason: "missing-target",
			failedAt: segment
		};
		return {
			kind: "ok",
			value: cursor[i]
		};
	}
	const obj = cursor;
	if (!Object.prototype.hasOwnProperty.call(obj, segment)) return {
		kind: "unresolved",
		reason: "missing-target",
		failedAt: segment
	};
	return {
		kind: "ok",
		value: obj[segment]
	};
};
const tryResolvePointer = (root, pointer) => {
	if (pointer === "#" || pointer === "") return {
		kind: "ok",
		value: root
	};
	if (!pointer.startsWith("#/")) return {
		kind: "unresolved",
		reason: "malformed-pointer"
	};
	const segments = pointer.slice(2).split("/").map(decodePointerSegment);
	let cursor = root;
	for (const seg of segments) {
		const step = tryStep(cursor, seg);
		if (step.kind === "unresolved") return step;
		cursor = step.value;
	}
	return {
		kind: "ok",
		value: cursor
	};
};
const throwResolutionError = (pointer, result) => {
	if (result.reason === "malformed-pointer") throw new JsonSchemaRefResolutionError(`Unsupported $ref pointer: ${pointer}`, {
		meta: { ref: pointer },
		possibleFixes: REF_RESOLUTION_FIXES
	});
	throw new JsonSchemaRefResolutionError(`Cannot resolve $ref ${pointer}`, {
		meta: {
			ref: pointer,
			...result.failedAt !== void 0 ? { failedAt: result.failedAt } : {}
		},
		possibleFixes: REF_RESOLUTION_FIXES
	});
};
/**
* Inlines internal JSON Schema `$ref` pointers (`#/$defs/...` and legacy
* `#/definitions/...`) so the returned schema can be safely handed to
* consumers that don't tolerate unresolved references (e.g. AJV in
* `@mastra/schema-compat`). External (`http://`, `https://`, …) refs are
* left untouched. Cycles are broken with `{ type: 'object',
* additionalProperties: true }`. The input is never mutated.
*
* By default, unresolved internal refs throw `JsonSchemaRefResolutionError`.
* Pass `{ onUnresolved: 'sentinel' }` to replace the offending node with the
* cycle-break sentinel instead — appropriate for schemas sourced from an
* upstream service the caller cannot edit (the Composio API ships some
* `outputParameters` with a `$ref` into `#/$defs/...` but never declares
* `$defs`; see https://github.com/ComposioHQ/composio/issues/3307).
*
* @throws {JsonSchemaRefResolutionError} on malformed pointers, missing
* targets (strict mode only), or chains past the depth cap (both modes).
*/
function dereferenceJsonSchema(schema, options) {
	if (!isPlainObject(schema)) return schema;
	const strategy = options?.onUnresolved ?? "throw";
	const onReplace = options?.onReplace;
	const root = schema;
	const visiting = /* @__PURE__ */ new WeakSet();
	const cloneChildren = (obj, visitedRefs, chainDepth, nodeDepth) => Object.fromEntries(Object.entries(obj).filter(([k]) => !POLLUTING_KEYS.has(k)).map(([k, v]) => [k, walk(v, visitedRefs, chainDepth, nodeDepth + 1)]));
	function walk(node, visitedRefs, chainDepth, nodeDepth) {
		if (nodeDepth >= MAX_NODE_DEPTH) throw new JsonSchemaRefResolutionError(`JSON Schema node depth exceeded cap (${MAX_NODE_DEPTH})`, { possibleFixes: REF_RESOLUTION_FIXES });
		if (Array.isArray(node)) {
			if (visiting.has(node)) return { ...CYCLE_BREAK_SENTINEL };
			visiting.add(node);
			try {
				return node.map((item) => walk(item, visitedRefs, chainDepth, nodeDepth + 1));
			} finally {
				visiting.delete(node);
			}
		}
		if (!isPlainObject(node)) return node;
		if (visiting.has(node)) return { ...CYCLE_BREAK_SENTINEL };
		visiting.add(node);
		try {
			const ref = typeof node.$ref === "string" ? node.$ref : null;
			if (ref === null || !ref.startsWith("#")) {
				if (ref !== null) logger.warn(`Leaving external $ref untouched: ${ref}`);
				return cloneChildren(node, visitedRefs, chainDepth, nodeDepth);
			}
			if (chainDepth >= MAX_REF_CHAIN_DEPTH) throw new JsonSchemaRefResolutionError(`JSON Schema $ref chain exceeded depth cap (${MAX_REF_CHAIN_DEPTH}): ${ref}`, {
				meta: { ref },
				possibleFixes: REF_RESOLUTION_FIXES
			});
			if (visitedRefs.has(ref)) return { ...CYCLE_BREAK_SENTINEL };
			const result = tryResolvePointer(root, ref);
			let target;
			if (result.kind === "ok") target = result.value;
			else if (strategy === "sentinel") {
				onReplace?.(ref, result.reason);
				target = {
					...CYCLE_BREAK_SENTINEL,
					description: UNRESOLVED_REF_DESCRIPTION
				};
			} else throwResolutionError(ref, result);
			const nextRefs = new Set(visitedRefs).add(ref);
			const resolved = walk(target, nextRefs, chainDepth + 1, nodeDepth + 1);
			const siblings = { ...node };
			delete siblings.$ref;
			if (Object.keys(siblings).length === 0 || !isPlainObject(resolved)) return resolved;
			return {
				...resolved,
				...cloneChildren(siblings, visitedRefs, chainDepth, nodeDepth)
			};
		} finally {
			visiting.delete(node);
		}
	}
	const out = walk(root, /* @__PURE__ */ new Set(), 0, 0);
	if (isPlainObject(out)) {
		delete out.$defs;
		delete out.definitions;
	}
	return out;
}
/**
* Returns a deep-cloned JSON Schema whose `required` arrays contain each entry
* at most once. Duplicate entries are invalid in JSON Schema 2020-12 but do
* not change the schema's meaning, so preserving the first occurrence is safe.
*
* Tool schemas originate both from the Composio API and direct provider calls.
* Keeping this normalization provider-agnostic lets every SDK provider receive
* canonical API schemas, while providers that further transform a schema can
* reapply it at their own emission boundary.
*/
function deduplicateJsonSchemaRequiredArrays(schema) {
	const seenSchemaValues = /* @__PURE__ */ new WeakMap();
	const seenInstanceValues = /* @__PURE__ */ new WeakMap();
	const instanceValueKeywords = /* @__PURE__ */ new Set([
		"const",
		"default",
		"enum",
		"examples"
	]);
	function walk(value, isSchema, depth = 0) {
		if (depth > MAX_NODE_DEPTH) throw new RangeError(`JSON Schema exceeds maximum nesting depth of ${MAX_NODE_DEPTH}`);
		const seen = isSchema ? seenSchemaValues : seenInstanceValues;
		if (Array.isArray(value)) {
			const existing = seen.get(value);
			if (existing) return existing;
			const clone = [];
			seen.set(value, clone);
			for (const item of value) clone.push(walk(item, isSchema, depth + 1));
			return clone;
		}
		if (!isPlainObject(value)) return value;
		const existing = seen.get(value);
		if (existing) return existing;
		const clone = {};
		seen.set(value, clone);
		for (const [key, child] of Object.entries(value)) Object.defineProperty(clone, key, {
			configurable: true,
			enumerable: true,
			value: isSchema && key === "required" && Array.isArray(child) ? [...new Set(walk(child, false, depth + 1))] : walk(child, isSchema && !instanceValueKeywords.has(key), depth + 1),
			writable: true
		});
		return clone;
	}
	return walk(schema, true);
}
/**
* Removes all non-required properties from the schema
*
* if no items are required, the schema is returned as is
* @param schema - The JSON schema to remove non-required properties from
* @returns The JSON schema with all non-required properties removed
*/
const removeNonRequiredProperties = (schema) => {
	if (schema && schema.type === "object" && schema.required?.length) schema.properties = Object.fromEntries(Object.entries(schema.properties || {}).filter(([key]) => schema.required.includes(key)));
	schema.additionalProperties = false;
	return schema;
};
/**
* Convert a JSON schema to a Zod schema
* @param jsonSchema - The JSON schema to convert
* @param strict - Eliminates all non-required properties from the schema
* @returns The Zod schema
*
* @throws {JsonSchemaToZodError} If the JSON schema is invalid
*
* @example
* ```ts
* const zodSchema = jsonSchemaToZodSchema({
*   type: 'object',
*   properties: {
*     name: { type: 'string' },
*   },
* });
*
* console.log(zodSchema);
* ```
*
* @example
* ```ts
* const zodSchema = jsonSchemaToZodSchema({
*   type: 'object',
*   properties: {
*     name: { type: 'string' },
*     age: { type: 'number' },
*   },
*   required: ['name'],
* }, { strict: true });
*
* console.log(zodSchema);
*
* // Output:
* // z.object({
* //   name: z.string(),
* // })
* ```
*/
function jsonSchemaToZodSchema(jsonSchema, { strict } = { strict: false }) {
	try {
		let schema = jsonSchema;
		if (strict && schema) schema = removeNonRequiredProperties(schema);
		return jsonSchemaToZod(schema);
	} catch (error) {
		throw new JsonSchemaToZodError("Failed to convert JSON Schema to Zod Schema", { cause: error });
	}
}

//#endregion
//#region src/utils/cancellation.ts
/** @internal */
async function withCancellation(call, signal) {
	try {
		return await call();
	} catch (error) {
		if (signal?.aborted && isRequestAbortError(error)) {
			const underlyingMessage = error instanceof Error ? error.message : "";
			throw new ComposioRequestCancelledError(underlyingMessage ? `Request was cancelled by the caller: ${underlyingMessage}` : "Request was cancelled by the caller", { cause: error instanceof Error ? error : void 0 });
		}
		throw error;
	}
}

//#endregion
export { removeNonRequiredProperties as a, jsonSchemaToZodSchema as i, deduplicateJsonSchemaRequiredArrays as n, dereferenceJsonSchema as r, withCancellation as t };
