// @composio/core/dist/CustomTool-C_5NvBsX.mjs
import { _ as ValidationError } from "./errors-B165yJUQ.mjs";
import { z as z$1 } from "zod/v3";
import * as z4 from "zod/v4";
import * as zodToJsonSchema from "zod-to-json-schema";

//#region src/types/customTool.types.ts
/**
* Zod schema for validating a custom tool slug.
* Alphanumeric, underscores, and hyphens only. No `LOCAL_` or `COMPOSIO_` prefix.
* Length is validated contextually (standalone vs extension vs toolkit) at creation time.
*/
const CustomToolSlugSchema = z$1.string().min(1, "slug is required").regex(/^[A-Za-z0-9_-]+$/, "slug must only contain alphanumeric characters, underscores, and hyphens").refine((s) => !s.toUpperCase().startsWith("LOCAL_"), { message: "slug must not start with \"LOCAL_\" — this prefix is reserved for internal routing." }).refine((s) => !s.toUpperCase().startsWith("COMPOSIO_"), { message: "slug must not start with \"COMPOSIO_\" — this prefix is reserved for Composio meta tools." });
/**
* Zod schema for validating the string/scalar fields of createCustomTool() options.
* Slug is validated separately as the first argument.
* Used internally for validation — inputParams, outputParams, and execute are checked manually.
*/
const CreateCustomToolBaseSchema = z$1.object({
	name: z$1.string().min(1, "createCustomTool: name is required"),
	description: z$1.string().min(1, "createCustomTool: description is required"),
	/**
	* Composio toolkit slug that this tool extends (inherits auth from).
	* Must be a valid Composio toolkit slug (e.g. `'gmail'`, `'github'`, `'meta_ads'`).
	* The backend validates this against the toolkit catalog.
	* Leave empty for tools that don't need any Composio-managed authentication.
	*/
	extendsToolkit: z$1.string().optional(),
	preload: z$1.boolean().optional().describe("When true, include this custom tool in session.tools() so it can be called without searching first.")
});
/**
* Zod schema for validating the string/scalar fields of createCustomToolkit() options.
* Slug is validated separately as the first argument.
*/
const CreateCustomToolkitBaseSchema = z$1.object({
	name: z$1.string().min(1, "createCustomToolkit: name is required"),
	description: z$1.string().min(1, "createCustomToolkit: description is required"),
	preload: z$1.boolean().optional().describe("When true, include tools from this custom toolkit in session.tools(). Tool-level preload values override this default.")
});

//#endregion
//#region src/utils/zodSchema.ts
const getZodDef = (value) => typeof value === "object" && value !== null ? value._def : void 0;
const isZodV4Schema = (schema) => typeof getZodDef(schema)?.type === "string";
const stripSchemaKeyword = (schema) => {
	const { $schema: _schema, ...rest } = schema;
	return rest;
};
const getNamedDefinition = (schema, name) => {
	const definition = schema.definitions?.[name];
	if (typeof definition !== "object" || definition === null) throw new Error(`Expected definition "${name}" in the converted JSON schema.`);
	return definition;
};
const isZodObjectSchema = (value) => {
	const def = getZodDef(value);
	return def?.typeName === "ZodObject" || def?.type === "object";
};
const zodSchemaToJsonSchema = (schema, name = "schema", mode = "output") => {
	if (isZodV4Schema(schema)) return stripSchemaKeyword(z4.toJSONSchema(schema, { io: mode }));
	const converted = zodToJsonSchema.default(schema, {
		name,
		pipeStrategy: mode,
		effectStrategy: mode === "input" ? "input" : "any"
	});
	return stripSchemaKeyword(getNamedDefinition(converted, name));
};
const zodObjectSchemaToJsonSchema = (schema, name = "schema", mode = "input") => {
	if (!isZodObjectSchema(schema)) throw new Error("Expected a z.object() schema.");
	const jsonSchema = zodSchemaToJsonSchema(schema, name, mode);
	return {
		type: "object",
		properties: jsonSchema.properties ?? {},
		...jsonSchema.required ? { required: jsonSchema.required } : {}
	};
};

//#endregion
//#region src/models/CustomTool.ts
/**
* @fileoverview Factory functions for creating custom tools and toolkits.
*
* Usage:
* ```typescript
* import { experimental_createTool, experimental_createToolkit } from '@composio/core';
*
* const grep = createCustomTool('GREP', {
*   name: 'Grep Search',
*   description: 'Search for patterns in files',
*   inputParams: z.object({ pattern: z.string() }),
*   execute: async (input) => ({ matches: [] }),
* });
*
* const devTools = createCustomToolkit('DEV_TOOLS', {
*   name: 'Dev Tools',
*   description: 'Local dev utilities',
*   tools: [grep],
* });
* ```
*/
/** Prefix applied by the backend to local tool slugs for disambiguation. */
const LOCAL_TOOL_PREFIX = "LOCAL_";
/** Maximum allowed length for the final prefixed slug. */
const MAX_SLUG_LENGTH = 60;
/**
* Compute the final slug length for a tool given its context.
* Returns the expected length of LOCAL_[TOOLKIT_]SLUG.
*/
function computeFinalSlugLength(toolSlug, toolkitSlug) {
	return 6 + (toolkitSlug ? toolkitSlug.length + 1 : 0) + toolSlug.length;
}
/**
* Validate that the final slug won't exceed the max length.
* Called early in createCustomTool/createCustomToolkit for fast feedback.
*/
function validateSlugLength(toolSlug, toolkitSlug, context) {
	const finalLength = computeFinalSlugLength(toolSlug, toolkitSlug);
	if (finalLength > MAX_SLUG_LENGTH) {
		const prefix = LOCAL_TOOL_PREFIX + (toolkitSlug ? `${toolkitSlug.toUpperCase()}_` : "");
		const available = MAX_SLUG_LENGTH - prefix.length;
		throw new ValidationError(`${context}: slug "${toolSlug}" is too long. With prefix "${prefix}", the final slug would be ${finalLength} characters (max ${MAX_SLUG_LENGTH}). Shorten the slug to at most ${available} characters.`);
	}
}
/**
* Create a custom tool for use in tool router sessions.
*
* The returned object is a lightweight reference containing the tool's metadata
* and execute function. Pass it to `composio.create(userId, { experimental: { customTools: [...] } })`
* to bind it to a session.
*
* Just return the result data from `execute`, or throw an error.
* The SDK wraps it into the standard response format internally.
*
* **Slug naming:** The slug you provide is automatically prefixed with `LOCAL_` when
* exposed to the agent (e.g. `'GREP'` becomes `LOCAL_GREP`). If the tool is inside a
* custom toolkit, the toolkit slug is also included (e.g. `LOCAL_DEV_TOOLS_GREP`).
* The `LOCAL_` prefix is reserved and cannot be used in your slug.
*
* @param slug - Unique tool identifier (alphanumeric, underscores, hyphens; no LOCAL_ or COMPOSIO_ prefix)
* @param options - Tool definition including name, schema, and execute function
* @returns A CustomTool to pass to session creation
*
* @example Standalone tool (no auth)
* ```typescript
* const grep = createCustomTool('GREP', {
*   name: 'Grep Search',
*   description: 'Search for patterns in files',
*   inputParams: z.object({ pattern: z.string(), path: z.string() }),
*   execute: async (input) => ({ matches: [] }),
* });
* ```
*
* @example Tool extending a Composio toolkit (inherits auth)
* ```typescript
* const getImportant = createCustomTool('GET_IMPORTANT_EMAILS', {
*   name: 'Get Important Emails',
*   description: 'Fetch high-priority emails',
*   extendsToolkit: 'gmail',
*   inputParams: z.object({ limit: z.number().default(10) }),
*   execute: async (input, ctx) => {
*     // Same response shape as session.execute(): { data, error, logId }
*     const result = await ctx.execute('GMAIL_SEARCH', { query: 'is:important' });
*     return { emails: result.data };
*   },
* });
* ```
*/
function createCustomTool(slug, options) {
	const slugResult = CustomToolSlugSchema.safeParse(slug);
	if (!slugResult.success) throw new ValidationError(`createCustomTool: ${slugResult.error.issues[0].message}`, { cause: slugResult.error });
	const validated = CreateCustomToolBaseSchema.parse(options);
	if (!options.inputParams) throw new ValidationError("createCustomTool: inputParams is required");
	if (!isZodObjectSchema(options.inputParams)) throw new ValidationError("createCustomTool: inputParams must be a z.object() schema. Tool input parameters are always an object with named properties.");
	if (typeof options.execute !== "function") throw new ValidationError("createCustomTool: execute must be a function");
	validateSlugLength(slug, validated.extendsToolkit, "createCustomTool");
	const { inputParams, execute } = options;
	const inputSchema = zodObjectSchemaToJsonSchema(inputParams, "input");
	let outputSchema;
	if (options.outputParams) outputSchema = zodSchemaToJsonSchema(options.outputParams, "output");
	return {
		slug: slugResult.data,
		name: validated.name,
		description: validated.description,
		extendsToolkit: validated.extendsToolkit,
		preload: validated.preload,
		inputSchema,
		outputSchema,
		inputParams,
		execute
	};
}
/**
* Create a custom toolkit that groups related tools.
*
* Tools passed here must NOT have `extendsToolkit` set — they inherit the toolkit identity instead.
*
* **Slug naming:** The toolkit slug becomes part of the final tool slug exposed to the agent.
* For example, a toolkit `'DEV_TOOLS'` with a tool `'GREP'` produces `LOCAL_DEV_TOOLS_GREP`.
* The `LOCAL_` prefix is reserved and cannot be used in your slug.
*
* @param slug - Unique toolkit identifier (alphanumeric, underscores, hyphens; no LOCAL_ or COMPOSIO_ prefix)
* @param options - Toolkit definition including name, description, and tools
* @returns A CustomToolkit to pass to session creation
*
* @example
* ```typescript
* const devTools = createCustomToolkit('DEV_TOOLS', {
*   name: 'Dev Tools',
*   description: 'Local dev utilities',
*   tools: [grepTool, sedTool],
* });
* ```
*/
function createCustomToolkit(slug, options) {
	const slugResult = CustomToolSlugSchema.safeParse(slug);
	if (!slugResult.success) throw new ValidationError(`createCustomToolkit: ${slugResult.error.issues[0].message}`, { cause: slugResult.error });
	const validated = CreateCustomToolkitBaseSchema.parse(options);
	if (!options.tools?.length) throw new ValidationError("createCustomToolkit: at least one tool is required");
	for (const tool of options.tools) {
		if (tool.extendsToolkit) throw new ValidationError(`createCustomToolkit: tool "${tool.slug}" has extendsToolkit set. Tools in a custom toolkit must not use extendsToolkit — they inherit the toolkit identity instead.`);
		validateSlugLength(tool.slug, slug, `createCustomToolkit("${slug}")`);
	}
	return {
		slug: slugResult.data,
		name: validated.name,
		description: validated.description,
		preload: validated.preload,
		tools: options.tools
	};
}
/**
* Build the final slug for a custom tool given its context.
* @internal
*/
function buildFinalSlug(toolSlug, toolkitSlug) {
	const upper = toolSlug.toUpperCase();
	return toolkitSlug ? `${LOCAL_TOOL_PREFIX}${toolkitSlug.toUpperCase()}_${upper}` : `${LOCAL_TOOL_PREFIX}${upper}`;
}
const qualifiedOriginalSlugKey = (toolkit, originalSlug) => `${toolkit?.toUpperCase() ?? ""}::${originalSlug.toUpperCase()}`;
const canShareOriginalSlug = (existing, next) => !!existing.toolkit && !!next.toolkit && existing.toolkit.toLowerCase() !== next.toolkit.toLowerCase();
const addOriginalSlugAlias = (params) => {
	const originalSlugKey = params.originalSlug.toUpperCase();
	if (params.ambiguousOriginalSlugs.has(originalSlugKey)) return;
	const existing = params.byOriginalSlug.get(originalSlugKey);
	if (!existing) {
		params.byOriginalSlug.set(originalSlugKey, params.entry);
		return;
	}
	if (existing.finalSlug.toUpperCase() === params.entry.finalSlug.toUpperCase()) return;
	if (canShareOriginalSlug(existing, params.entry)) {
		params.byOriginalSlug.delete(originalSlugKey);
		params.ambiguousOriginalSlugs.add(originalSlugKey);
		return;
	}
	throw new ValidationError(`Custom tool slug collision: original slug "${params.originalSlug}" maps to multiple final slugs. "${existing.finalSlug}" and "${params.entry.finalSlug}" both resolve from "${originalSlugKey}".`);
};
/**
* Build a CustomToolsMap from custom tools and toolkits.
* Used internally by ToolRouter.create() to construct the per-session routing map.
*
* @internal
* @param tools - Standalone custom tools
* @param toolkits - Custom toolkits containing grouped tools
* @returns Maps for O(1) lookup by both final and original slug
* @throws If duplicate slugs or cross-group collisions
*/
function buildCustomToolsMap(tools, toolkits) {
	const byFinalSlug = /* @__PURE__ */ new Map();
	const byOriginalSlug = /* @__PURE__ */ new Map();
	const byToolkitAndOriginalSlug = /* @__PURE__ */ new Map();
	const ambiguousOriginalSlugs = /* @__PURE__ */ new Set();
	const addEntry = (handle, finalSlug, toolkit) => {
		const originalSlug = handle.slug.toUpperCase();
		const finalSlugKey = finalSlug.toUpperCase();
		if (finalSlug.length > MAX_SLUG_LENGTH) throw new ValidationError(`Custom tool slug "${handle.slug}" produces final slug "${finalSlug}" which exceeds ${MAX_SLUG_LENGTH} characters.`);
		if (byFinalSlug.has(finalSlugKey)) throw new ValidationError(`Custom tool slug collision: "${finalSlug}" is already registered.`);
		const qualifiedKey = qualifiedOriginalSlugKey(toolkit, originalSlug);
		if (byToolkitAndOriginalSlug.has(qualifiedKey)) throw new ValidationError(`Custom tool slug collision: original slug "${handle.slug}" is already registered for toolkit "${toolkit ?? "custom"}".`);
		const entry = {
			handle,
			finalSlug,
			toolkit
		};
		byFinalSlug.set(finalSlugKey, entry);
		byToolkitAndOriginalSlug.set(qualifiedKey, entry);
		addOriginalSlugAlias({
			byOriginalSlug,
			ambiguousOriginalSlugs,
			originalSlug,
			entry
		});
	};
	for (const handle of tools) addEntry(handle, buildFinalSlug(handle.slug, handle.extendsToolkit), handle.extendsToolkit);
	if (toolkits) for (const toolkit of toolkits) for (const handle of toolkit.tools) addEntry(handle, buildFinalSlug(handle.slug, toolkit.slug), toolkit.slug);
	return {
		byFinalSlug,
		byOriginalSlug,
		byToolkitAndOriginalSlug,
		...ambiguousOriginalSlugs.size ? { ambiguousOriginalSlugs } : {},
		toolkits,
		tools: tools.length ? tools : void 0
	};
}
function shouldSerializePreload(preload, inheritedPreload, defaultPreload) {
	const inheritedOrDefault = inheritedPreload ?? defaultPreload;
	if (preload !== void 0) return preload || inheritedOrDefault ? preload : void 0;
	if (inheritedPreload !== void 0) return inheritedPreload || defaultPreload ? inheritedPreload : void 0;
	return defaultPreload ? true : void 0;
}
function preloadWireProperty(preload, inheritedPreload, defaultPreload) {
	const serializedPreload = shouldSerializePreload(preload, inheritedPreload, defaultPreload);
	return serializedPreload === void 0 ? {} : { preload: serializedPreload };
}
function serializeCustomTools(tools, options = {}) {
	const defaultPreload = options.defaultPreload ?? false;
	return tools.map((handle) => ({
		slug: handle.slug,
		name: handle.name,
		description: handle.description,
		input_schema: handle.inputSchema,
		...handle.outputSchema ? { output_schema: handle.outputSchema } : {},
		...handle.extendsToolkit ? { extends_toolkit: handle.extendsToolkit } : {},
		...preloadWireProperty(handle.preload, void 0, defaultPreload)
	}));
}
/**
* Serialize custom toolkits into the format expected by the backend session creation API.
*
* @internal
* @param toolkits - The custom toolkits to serialize
* @returns Array of CustomToolkitDefinition for the API payload
*/
function serializeCustomToolkits(toolkits, options = {}) {
	const defaultPreload = options.defaultPreload ?? false;
	return toolkits.map((tk) => ({
		slug: tk.slug,
		name: tk.name,
		description: tk.description,
		...preloadWireProperty(tk.preload, void 0, defaultPreload),
		tools: tk.tools.map((t) => ({
			slug: t.slug,
			name: t.name,
			description: t.description,
			input_schema: t.inputSchema,
			...t.outputSchema ? { output_schema: t.outputSchema } : {},
			...preloadWireProperty(t.preload, tk.preload, defaultPreload)
		}))
	}));
}
/**
* Build a CustomToolsMap using the slug/original_slug mapping from the session create response.
* This uses the backend's authoritative prefixed slugs instead of computing them client-side.
*
* @internal
* @param tools - The original custom tools passed to session creation
* @param toolkits - The original custom toolkits passed to session creation
* @param experimental - The experimental section from the session create response
* @returns Maps for O(1) lookup by both final and original slug
*/
function buildCustomToolsMapFromResponse(tools, toolkits, experimental) {
	const byFinalSlug = /* @__PURE__ */ new Map();
	const byOriginalSlug = /* @__PURE__ */ new Map();
	const byToolkitAndOriginalSlug = /* @__PURE__ */ new Map();
	const ambiguousOriginalSlugs = /* @__PURE__ */ new Set();
	const handlesByQualifiedOriginalSlug = /* @__PURE__ */ new Map();
	const handlesByOriginalSlug = /* @__PURE__ */ new Map();
	const registerHandle = (handle, toolkit) => {
		const originalSlug = handle.slug.toUpperCase();
		const match = {
			handle,
			toolkit
		};
		handlesByQualifiedOriginalSlug.set(qualifiedOriginalSlugKey(toolkit, originalSlug), match);
		const existing = handlesByOriginalSlug.get(originalSlug) ?? [];
		existing.push(match);
		handlesByOriginalSlug.set(originalSlug, existing);
	};
	for (const handle of tools) registerHandle(handle, handle.extendsToolkit);
	if (toolkits) for (const tk of toolkits) for (const handle of tk.tools) registerHandle(handle, tk.slug);
	const findHandle = (originalSlug, toolkit) => {
		const originalSlugKey = originalSlug.toUpperCase();
		const qualifiedMatch = handlesByQualifiedOriginalSlug.get(qualifiedOriginalSlugKey(toolkit, originalSlugKey));
		if (qualifiedMatch) return qualifiedMatch;
		const bareMatches = handlesByOriginalSlug.get(originalSlugKey) ?? [];
		return bareMatches.length === 1 ? bareMatches[0] : void 0;
	};
	const addEntry = (finalSlug, originalSlug, toolkit) => {
		const resolvedToolkit = toolkit ?? void 0;
		const match = findHandle(originalSlug, resolvedToolkit);
		if (!match) return;
		const finalSlugKey = finalSlug.toUpperCase();
		if (byFinalSlug.has(finalSlugKey)) throw new ValidationError(`Custom tool slug collision: "${finalSlug}" is already registered.`);
		const entry = {
			handle: match.handle,
			finalSlug,
			toolkit: resolvedToolkit ?? match.toolkit
		};
		byFinalSlug.set(finalSlugKey, entry);
		byToolkitAndOriginalSlug.set(qualifiedOriginalSlugKey(entry.toolkit, originalSlug), entry);
		addOriginalSlugAlias({
			byOriginalSlug,
			ambiguousOriginalSlugs,
			originalSlug,
			entry
		});
	};
	if (experimental?.custom_tools) for (const ct of experimental.custom_tools) addEntry(ct.slug, ct.original_slug, ct.extends_toolkit);
	if (experimental?.custom_toolkits) for (const ctk of experimental.custom_toolkits) for (const ct of ctk.tools) addEntry(ct.slug, ct.original_slug, ctk.slug);
	return {
		byFinalSlug,
		byOriginalSlug,
		byToolkitAndOriginalSlug,
		...ambiguousOriginalSlugs.size ? { ambiguousOriginalSlugs } : {},
		toolkits,
		tools: tools.length ? tools : void 0
	};
}
/**
* Find a custom tool entry by final slug only.
*
* @internal
*/
function findCustomToolMapEntryByFinalSlug(customToolsMap, slug) {
	return customToolsMap?.byFinalSlug.get(slug.toUpperCase());
}
/**
* Find a custom toolkit tool entry by toolkit slug and original tool slug.
*
* @internal
*/
function findCustomToolMapEntryByToolkitAndOriginalSlug(customToolsMap, toolkit, slug) {
	return customToolsMap?.byToolkitAndOriginalSlug?.get(qualifiedOriginalSlugKey(toolkit, slug));
}
/**
* Reject legacy attempts to preload custom tools through top-level preload.tools.
*
* Custom tool direct exposure is controlled by CustomTool.preload /
* CustomToolkit.preload so it follows the current SDK-provided custom
* definitions instead of stale session config.
*
* @internal
*/
function assertNoCustomToolSlugsInPreload(preloadTools, customToolsMap) {
	if (!preloadTools || preloadTools === "all") return;
	const customPreloadSlugs = preloadTools.filter((slug) => {
		const normalized = slug.toUpperCase();
		return normalized.startsWith("LOCAL_") || customToolsMap?.byOriginalSlug.has(normalized) || customToolsMap?.ambiguousOriginalSlugs?.has(normalized) || customToolsMap?.byFinalSlug.has(normalized);
	});
	if (customPreloadSlugs.length) throw new ValidationError(`Custom tool slugs are not supported in preload.tools: ${customPreloadSlugs.join(", ")}. Set preload: true on the SDK custom tool or custom toolkit definition instead.`);
}
/**
* Resolve custom tools that the SDK should expose directly from session.tools().
*
* @internal
*/
function getPreloadedCustomToolSlugs(customToolsMap, defaultPreload = false) {
	if (!customToolsMap) return [];
	const seen = /* @__PURE__ */ new Set();
	const customToolSlugs = [];
	for (const entry of customToolsMap.byFinalSlug.values()) {
		const toolkit = customToolsMap.toolkits?.find((tk) => entry.toolkit && tk.slug.toLowerCase() === entry.toolkit.toLowerCase());
		if (!(entry.handle.preload ?? toolkit?.preload ?? defaultPreload)) continue;
		const finalSlugKey = entry.finalSlug.toUpperCase();
		if (seen.has(finalSlugKey)) continue;
		seen.add(finalSlugKey);
		customToolSlugs.push(entry.finalSlug);
	}
	return customToolSlugs;
}

//#endregion
export { createCustomToolkit as a, getPreloadedCustomToolSlugs as c, CreateCustomToolBaseSchema as d, CreateCustomToolkitBaseSchema as f, createCustomTool as i, serializeCustomToolkits as l, buildCustomToolsMap as n, findCustomToolMapEntryByFinalSlug as o, CustomToolSlugSchema as p, buildCustomToolsMapFromResponse as r, findCustomToolMapEntryByToolkitAndOriginalSlug as s, assertNoCustomToolSlugsInPreload as t, serializeCustomTools as u };
