// @composio/core/dist/errors-B165yJUQ.mjs
import { l as ComposioError$1 } from "./SsrfErrors-BQRozkeU.mjs";
import { APIError, APIUserAbortError } from "@composio/client";
import { ZodError } from "zod/v3";

//#region src/errors/AuthConfigErrors.ts
const AuthConfigErrorCodes = { AUTH_CONFIG_NOT_FOUND: "AUTH_CONFIG_NOT_FOUND" };
var ComposioAuthConfigNotFoundError = class extends ComposioError$1 {
	constructor(message = "Auth config not found", options = {}) {
		super(message, {
			...options,
			code: AuthConfigErrorCodes.AUTH_CONFIG_NOT_FOUND,
			possibleFixes: options.possibleFixes || [
				"Check if the auth config exists",
				"Check if the auth config id is correct",
				"Check if the auth config is enabled"
			]
		});
		this.name = "ComposioAuthConfigNotFoundError";
	}
};

//#endregion
//#region src/errors/ConnectedAccountsErrors.ts
const ConnectedAccountErrorCodes = {
	CONNECTED_ACCOUNT_NOT_FOUND: "CONNECTED_ACCOUNT_NOT_FOUND",
	MULTIPLE_CONNECTED_ACCOUNTS: "MULTIPLE_CONNECTED_ACCOUNTS",
	FAILED_TO_CREATE_CONNECTED_ACCOUNT_LINK: "FAILED_TO_CREATE_CONNECTED_ACCOUNT_LINK",
	LEGACY_CONNECTED_ACCOUNTS_ENDPOINT_RETIRED: "LEGACY_CONNECTED_ACCOUNTS_ENDPOINT_RETIRED",
	SHARED_ACCESS_DENIED: "SHARED_ACCESS_DENIED",
	ACL_ONLY_FOR_SHARED: "ACL_ONLY_FOR_SHARED",
	SHARED_CONNECTION_NOT_ACCESSIBLE: "SHARED_CONNECTION_NOT_ACCESSIBLE"
};
var ComposioConnectedAccountNotFoundError = class extends ComposioError$1 {
	constructor(message = "Connected account not found", options = {}) {
		super(message, {
			...options,
			code: ConnectedAccountErrorCodes.CONNECTED_ACCOUNT_NOT_FOUND,
			statusCode: 404,
			possibleFixes: options.possibleFixes || ["Ensure the connected account exists and is active in your Composio dashboard"]
		});
		this.name = "ComposioConnectedAccountNotFoundError";
	}
};
var ComposioMultipleConnectedAccountsError = class extends ComposioError$1 {
	constructor(message = "Multiple connected accounts found", options = {}) {
		super(message, {
			...options,
			code: ConnectedAccountErrorCodes.MULTIPLE_CONNECTED_ACCOUNTS,
			possibleFixes: options.possibleFixes || ["Use the allowMultiple flag to allow multiple connected accounts per user for an auth config"]
		});
		this.name = "ComposioMultipleConnectedAccountsError";
	}
};
var ComposioFailedToCreateConnectedAccountLink = class extends ComposioError$1 {
	constructor(message = "Failed to create connected account link", options = {}) {
		super(message, {
			...options,
			code: ConnectedAccountErrorCodes.FAILED_TO_CREATE_CONNECTED_ACCOUNT_LINK
		});
		this.name = "ComposioFailedToCreateConnectedAccountLink";
	}
};
/**
* Thrown when a tool execution attempts to use a SHARED connected account
* but the requesting `userId` is not allowed by the connection's ACL.
*
* Surfaces when a SHARED connection is reached directly (e.g. via
* `composio.tools.execute(slug, { connectedAccountId })`) without going
* through a tool-router session.
*/
var ComposioSharedAccessDeniedError = class extends ComposioError$1 {
	constructor(message = "Access denied: this SHARED connected account is not available to the requesting user", options = {}) {
		super(message, {
			...options,
			code: ConnectedAccountErrorCodes.SHARED_ACCESS_DENIED,
			statusCode: 403,
			possibleFixes: options.possibleFixes || ["Ask the connection's creator to grant access via composio.connectedAccounts.updateAcl() — set allowAllUsers, add the userId to allowedUserIds, or remove it from notAllowedUserIds."]
		});
		this.name = "ComposioSharedAccessDeniedError";
	}
};
/**
* Thrown when ACL fields (`allowAllUsers`, `allowedUserIds`,
* `notAllowedUserIds`) are sent on a `PRIVATE` connection — at create
* time or via PATCH. ACL is only meaningful for `SHARED` connections.
*/
var ComposioAclOnlyForSharedError = class extends ComposioError$1 {
	constructor(message = "ACL fields are only valid on SHARED connected accounts", options = {}) {
		super(message, {
			...options,
			code: ConnectedAccountErrorCodes.ACL_ONLY_FOR_SHARED,
			statusCode: 400,
			possibleFixes: options.possibleFixes || ["Use accountType: \"SHARED\", or omit ACL fields when creating/updating a PRIVATE connection."]
		});
		this.name = "ComposioAclOnlyForSharedError";
	}
};
/**
* Thrown by `toolRouter.session.create()` / `session.patch()` when the
* session's `userId` cannot use a pinned SHARED connection. Raised at
* session-create time so the session never enters a state that fails
* mid-execution.
*/
var ComposioSharedConnectionNotAccessibleError = class extends ComposioError$1 {
	constructor(message = "A SHARED connected account pinned in the session config is not accessible to the session user", options = {}) {
		super(message, {
			...options,
			code: ConnectedAccountErrorCodes.SHARED_CONNECTION_NOT_ACCESSIBLE,
			statusCode: 400,
			possibleFixes: options.possibleFixes || ["Grant the session user access via composio.connectedAccounts.updateAcl() on the pinned connection, or pin a different connection the user can use."]
		});
		this.name = "ComposioSharedConnectionNotAccessibleError";
	}
};
/**
* Thrown by `composio.connectedAccounts.initiate()` when the legacy
* `POST /api/v3/connected_accounts` endpoint rejects a Composio-managed
* OAuth (OAuth1, OAuth2, DCR_OAUTH) auth-config request.
*
* The retiring path is being phased out — new orgs from 2026-05-08 and
* all remaining orgs from 2026-07-03. Migrate the call to
* `composio.connectedAccounts.link()`, which works for every redirectable
* scheme regardless of whether the auth config is Composio-managed or
* custom.
*
* See: https://docs.composio.dev/docs/changelog/2026/04/24
*/
var ComposioLegacyConnectedAccountsEndpointRetiredError = class extends ComposioError$1 {
	constructor(message = "POST /api/v3/connected_accounts is no longer supported for Composio-managed OAuth auth configs. Use composio.connectedAccounts.link() instead.", options = {}) {
		super(message, {
			...options,
			code: ConnectedAccountErrorCodes.LEGACY_CONNECTED_ACCOUNTS_ENDPOINT_RETIRED,
			statusCode: 400,
			possibleFixes: options.possibleFixes || ["Replace `composio.connectedAccounts.initiate(userId, authConfigId, options)` with `composio.connectedAccounts.link(userId, authConfigId, options)`. The return shape (id, redirectUrl, waitForConnection()) is the same.", "Migration guide: https://docs.composio.dev/docs/changelog/2026/04/24"]
		});
		this.name = "ComposioLegacyConnectedAccountsEndpointRetiredError";
	}
};

//#endregion
//#region src/errors/ConnectionRequestErrors.ts
const ConnectionRequestErrorCodes = {
	CONNECTION_REQUEST_TIMEOUT: "CONNECTION_REQUEST_TIMEOUT",
	CONNECTION_REQUEST_FAILED: "CONNECTION_REQUEST_FAILED"
};
var ConnectionRequestTimeoutError = class extends ComposioError$1 {
	constructor(message = "Connection request timed out", options = {}) {
		super(message, {
			...options,
			code: ConnectionRequestErrorCodes.CONNECTION_REQUEST_TIMEOUT
		});
		this.name = "ConnectionRequestTimeoutError";
	}
};
var ConnectionRequestFailedError = class extends ComposioError$1 {
	constructor(message = "Connection request failed", options = {}) {
		super(message, {
			...options,
			code: ConnectionRequestErrorCodes.CONNECTION_REQUEST_FAILED
		});
		this.name = "ConnectionRequestFailedError";
	}
};

//#endregion
//#region src/errors/ToolErrors.ts
const ToolErrorCodes = {
	TOOLSET_NOT_DEFINED: "TOOLSET_NOT_DEFINED",
	TOOL_NOT_FOUND: "TOOL_NOT_FOUND",
	INVALID_MODIFIER: "INVALID_MODIFIER",
	TOOL_EXECUTION_ERROR: "TOOL_EXECUTION_ERROR",
	INVALID_TOOL_ARGUMENTS: "INVALID_TOOL_ARGUMENTS",
	INVALID_EXECUTE_FUNCTION: "INVALID_EXECUTE_FUNCTION",
	GLOBAL_EXECUTE_TOOL_FN_NOT_SET: "GLOBAL_EXECUTE_TOOL_FN_NOT_SET",
	TOOL_VERSION_REQUIRED: "TOOL_VERSION_REQUIRED"
};
var ComposioProviderNotDefinedError = class extends ComposioError$1 {
	constructor(message = "Provider not defined", options = {}) {
		super(message, {
			...options,
			code: ToolErrorCodes.TOOLSET_NOT_DEFINED,
			possibleFixes: options.possibleFixes || ["Ensure that the provider is defined in the Composio project and passed into the tool instance"]
		});
		this.name = "ComposioProviderNotDefinedError";
	}
};
var ComposioToolNotFoundError = class extends ComposioError$1 {
	constructor(message = "Tool not found", options = {}) {
		super(message, {
			...options,
			code: ToolErrorCodes.TOOL_NOT_FOUND,
			possibleFixes: options.possibleFixes || ["Ensure the tool slug is correct and exists in the Composio project"]
		});
		this.name = "ComposioToolNotFoundError";
	}
};
var ComposioInvalidModifierError = class extends ComposioError$1 {
	constructor(message = "Invalid modifier", options = {}) {
		super(message, {
			...options,
			code: ToolErrorCodes.INVALID_MODIFIER,
			possibleFixes: options.possibleFixes || ["Ensure the modifier is a function and returns a valid result"]
		});
		this.name = "ComposioInvalidModifierError";
	}
};
var ComposioToolExecutionError = class extends ComposioError$1 {
	constructor(message = "Tool execution error", options = {}) {
		super(message, {
			...options,
			code: options.code || ToolErrorCodes.TOOL_EXECUTION_ERROR,
			cause: options.cause,
			possibleFixes: options.possibleFixes || ["Ensure the tool is correctly configured and the input is valid", "Ensure the userId is correct and has an active connected account for the user in case of non NoAuth toolkits"]
		});
		this.name = "ComposioToolExecutionError";
	}
};
var ComposioInvalidToolArgumentsError = class extends ComposioError$1 {
	constructor(message = "Invalid tool arguments", options = {}) {
		super(message, {
			...options,
			code: ToolErrorCodes.INVALID_TOOL_ARGUMENTS,
			possibleFixes: options.possibleFixes || ["Tool arguments must be a JSON object (or a string that parses to one)", "If a model emitted the arguments as a string, ensure it is valid JSON"]
		});
		this.name = "ComposioInvalidToolArgumentsError";
	}
};
var ComposioInvalidExecuteFunctionError = class extends ComposioError$1 {
	constructor(message = "Invalid execute function", options = {}) {
		super(message, {
			...options,
			code: ToolErrorCodes.INVALID_EXECUTE_FUNCTION,
			possibleFixes: options.possibleFixes || ["Ensure the execute function is a valid function and returns a valid result"]
		});
		this.name = "ComposioInvalidExecuteFunctionError";
	}
};
var ComposioGlobalExecuteToolFnNotSetError = class extends ComposioError$1 {
	constructor(message = "Global execute tool function not set", options = {}) {
		super(message, {
			...options,
			code: ToolErrorCodes.GLOBAL_EXECUTE_TOOL_FN_NOT_SET,
			possibleFixes: options.possibleFixes || ["Ensure the global execute tool function is set in the provider"]
		});
		this.name = "ComposioGlobalExecuteToolFnNotSetError";
	}
};
/**
* Error thrown when toolkit version is not specified for manual tool execution.
*
* This error is thrown when attempting to execute a tool manually (via `tools.execute()`)
* and the resolved version is "latest" without `dangerouslySkipVersionCheck` set to true.
*
* **Why this error exists:**
* Using "latest" version in manual execution can lead to unexpected behavior when new
* toolkit versions are released, potentially breaking your application. For production
* use, it's recommended to pin specific toolkit versions.
*
* **How to fix:**
* 1. Specify a concrete version in the execute call
* 2. Configure toolkit versions at SDK initialization
* 3. Set toolkit version via environment variable
* 4. Use dangerouslySkipVersionCheck (not recommended for production)
*
* @example ❌ This will throw ComposioToolVersionRequiredError if toolkitVersions is 'latest'
* ```typescript
* await composio.tools.execute('GITHUB_GET_REPOS', {
*   userId: 'default',
*   arguments: { owner: 'composio' }
* });
* ```
*
* @example ✅ Fix 1: Pass specific version in execute call
* ```typescript
* await composio.tools.execute('GITHUB_GET_REPOS', {
*   userId: 'default',
*   version: '20250909_00',
*   arguments: { owner: 'composio' }
* });
* ```
*
* @example ✅ Fix 2: Configure toolkit versions at SDK level
* ```typescript
* const composio = new Composio({
*   toolkitVersions: { github: '20250909_00' }
* });
* await composio.tools.execute('GITHUB_GET_REPOS', {
*   userId: 'default',
*   arguments: { owner: 'composio' }
* });
* ```
*
* @example ✅ Fix 3: Use environment variable
* ```typescript
* // Set: COMPOSIO_TOOLKIT_VERSION_GITHUB=20250909_00
* await composio.tools.execute('GITHUB_GET_REPOS', {
*   userId: 'default',
*   arguments: { owner: 'composio' }
* });
* ```
*
* @example ⚠️ Fix 4: Skip version check (not recommended for production)
* ```typescript
* await composio.tools.execute('GITHUB_GET_REPOS', {
*   userId: 'default',
*   dangerouslySkipVersionCheck: true,
*   arguments: { owner: 'composio' }
* });
* ```
*/
var ComposioToolVersionRequiredError = class extends ComposioError$1 {
	constructor(message = "Toolkit version not specified. For manual execution of the tool please pass a specific toolkit version", options = {}) {
		super(message, {
			...options,
			code: ToolErrorCodes.TOOL_VERSION_REQUIRED,
			possibleFixes: options.possibleFixes || [
				"Pass the toolkit version as a parameter to the execute function (\"latest\" is not supported in manual execution)",
				"Set the toolkit versions in the Composio config (toolkitVersions: { <toolkit-slug>: \"<toolkit-version>\" })",
				"Set the toolkit version in the environment variable (COMPOSIO_TOOLKIT_VERSION_<TOOLKIT_SLUG>)",
				"Set dangerouslySkipVersionCheck to true (this might cause unexpected behavior when new versions of the tools are released)"
			]
		});
	}
};
const ERROR_CODE_HANDLERS = /* @__PURE__ */ new Map([[1803, (msg) => new ComposioConnectedAccountNotFoundError(msg)]]);
const handleToolExecutionError = (tool, actualError) => {
	if (actualError instanceof APIError && actualError.error) {
		const errorBody = actualError.error;
		const errorCode = errorBody?.error?.code;
		const errorMessage = errorBody?.error?.message;
		if (errorCode && ERROR_CODE_HANDLERS.has(errorCode)) return ERROR_CODE_HANDLERS.get(errorCode)(errorMessage || "An error occurred");
	}
	return new ComposioToolExecutionError(`Error executing the tool ${tool}`, {
		cause: actualError,
		possibleFixes: ["Ensure the tool slug is correct and the input arguments for the tool is valid"]
	});
};

//#endregion
//#region src/errors/ToolkitErrors.ts
const ToolkitErrorCodes = { TOOLKIT_NOT_FOUND: "TOOLKIT_NOT_FOUND" };
var ComposioToolkitNotFoundError = class extends ComposioError$1 {
	constructor(message = "Toolkit not found", options = {}) {
		super(message, {
			...options,
			code: "TOOLKIT_NOT_FOUND",
			possibleFixes: options.possibleFixes || ["Ensure the toolkit is correctly configured and the slug is valid"]
		});
		this.name = "ComposioToolkitNotFoundError";
	}
};
var ComposioToolkitFetchError = class extends ComposioError$1 {
	constructor(message = "Failed to fetch toolkit", options = {}) {
		super(message, {
			...options,
			code: "TOOLKIT_FETCH_ERROR",
			possibleFixes: options.possibleFixes || [
				"Ensure the toolkit slug is valid",
				"Ensure you are using the correct API key",
				"Ensure you are using the correct API endpoint / Base URL and it is working"
			]
		});
		this.name = "ComposioToolkitFetchError";
	}
};

//#endregion
//#region src/errors/ValidationErrors.ts
/**
* Represents an input validation error using Zod.
*/
const ValidationErrorCodes = {
	VALIDATION_ERROR: "VALIDATION_ERROR",
	JSON_SCHEMA_TO_ZOD_ERROR: "JSON_SCHEMA_TO_ZOD_ERROR",
	JSON_SCHEMA_REF_RESOLUTION_ERROR: "JSON_SCHEMA_REF_RESOLUTION_ERROR"
};
var ValidationError = class extends ComposioError$1 {
	constructor(message = "Input validation failed", options = {}) {
		const { cause: providedZodError, ...restOptions } = options;
		let zodErrorInstance;
		if (providedZodError instanceof ZodError) zodErrorInstance = providedZodError;
		else zodErrorInstance = new ZodError([{
			path: [],
			message: "Invalid input",
			code: "custom"
		}]);
		const issues = zodErrorInstance.issues.map((issue) => `[${issue.code}] ${issue.path.join(".")} - ${issue.message}`);
		super(message, {
			...restOptions,
			code: options.code || ValidationErrorCodes.VALIDATION_ERROR,
			possibleFixes: issues,
			cause: zodErrorInstance
		});
		this.name = "ValidationError";
		const userFriendlyMessage = this.generateUserFriendlyMessage();
		this.message = `${message}: ${userFriendlyMessage}`;
	}
	generateUserFriendlyMessage() {
		if (this.cause instanceof ZodError && this.cause.issues.length > 0) {
			const issue = this.cause.issues[0];
			const param = issue.path.join(".") || "parameter";
			if (issue.code === "invalid_type") return `The ${param} should be a ${issue.expected}, but you provided a ${issue.received}`;
			return issue.message;
		}
		return "Please check your input parameters";
	}
};
var JsonSchemaToZodError = class extends ComposioError$1 {
	constructor(message = "Failed to convert JSON schema to Zod schema", options = {}) {
		super(message, {
			...options,
			code: options.code || ValidationErrorCodes.JSON_SCHEMA_TO_ZOD_ERROR
		});
	}
};
/**
* Thrown when a JSON Schema `$ref` pointer cannot be resolved against the
* document. Used by `dereferenceJsonSchema` to surface malformed pointers,
* missing `$defs`/`definitions` targets, or chains exceeding the depth cap.
*/
var JsonSchemaRefResolutionError = class extends ComposioError$1 {
	constructor(message = "Failed to resolve JSON Schema $ref", options = {}) {
		super(message, {
			...options,
			code: options.code || ValidationErrorCodes.JSON_SCHEMA_REF_RESOLUTION_ERROR
		});
	}
};

//#endregion
//#region src/errors/SDKErrors.ts
const SDKErrorCodes = {
	NO_API_KEY_PROVIDED: "NO_API_KEY_PROVIDED",
	REQUEST_CANCELLED: "REQUEST_CANCELLED"
};
var ComposioNoAPIKeyError = class extends ComposioError$1 {
	constructor(message = "No Composio API key provided", options = {}) {
		const defaultCause = "Couldn't find an API key in the params, environment variables or in the user config file";
		super(message, {
			...options,
			code: SDKErrorCodes.NO_API_KEY_PROVIDED,
			cause: options.cause || defaultCause,
			possibleFixes: options.possibleFixes || [
				"Ensure you have an API key passed in the params, or in environment variable (COMPOSIO_API_KEY) or in the user config file",
				"To get an API key, please sign up at https://composio.dev/signup",
				"You can also use the Composio CLI to create a project and get an API key"
			],
			statusCode: 401
		});
		this.name = "ComposioNoAPIKeyError";
	}
};
/**
* Thrown when an SDK call is cancelled via the caller-supplied `AbortSignal`.
*
* @example
* ```typescript
* try {
*   await composio.tools.execute(slug, body, { signal: AbortSignal.timeout(5_000) });
* } catch (err) {
*   if (err instanceof ComposioRequestCancelledError) return;
*   throw err;
* }
* ```
*/
var ComposioRequestCancelledError = class extends ComposioError$1 {
	constructor(message = "Request was cancelled by the caller", options = {}) {
		super(message, {
			...options,
			code: SDKErrorCodes.REQUEST_CANCELLED,
			possibleFixes: options.possibleFixes ?? ["This error is expected when you abort the request via AbortController. Handle it in your catch block to distinguish caller-initiated cancellation from other failures."]
		});
		this.name = "ComposioRequestCancelledError";
	}
};
/**
* `APIUserAbortError` has `.name === "Error"` at runtime (doesn't override it),
* so we fall back to `constructor.name` for dual-package-hazard cases.
* @internal
*/
function isRequestAbortError(error) {
	return _isRequestAbortErrorAt(error, 0);
}
function _isRequestAbortErrorAt(error, depth) {
	if (depth > 5) return false;
	if (error instanceof APIUserAbortError) return true;
	if (typeof DOMException !== "undefined" && error instanceof DOMException && error.name === "AbortError") return true;
	if (!(error instanceof Error)) return false;
	if (error.constructor.name === "APIUserAbortError" || error.name === "AbortError" || error.name === "APIUserAbortError") return true;
	if ("cause" in error && error.cause !== void 0 && error.cause !== null) return _isRequestAbortErrorAt(error.cause, depth + 1);
	return false;
}

//#endregion
//#region src/errors/TriggerErrors.ts
const TriggerErrorCodes = {
	TRIGGER_FAILED_TO_GET_SDK_REALTIME_CREDENTIALS: "TRIGGER_FAILED_TO_GET_SDK_REALTIME_CREDENTIALS",
	TRIGGER_FAILED_TO_CREATE_PUSHER_CLIENT: "TRIGGER_FAILED_TO_CREATE_PUSHER_CLIENT",
	TRIGGER_FAILED_TO_SUBSCRIBE_TO_PUSHER_CHANNEL: "TRIGGER_FAILED_TO_SUBSCRIBE_TO_PUSHER_CHANNEL",
	TRIGGER_FAILED_TO_UNSUBSCRIBE_FROM_PUSHER_CHANNEL: "TRIGGER_FAILED_TO_UNSUBSCRIBE_FROM_PUSHER_CHANNEL",
	TRIGGER_TYPE_NOT_FOUND: "TRIGGER_TYPE_NOT_FOUND",
	WEBHOOK_SIGNATURE_VERIFICATION_FAILED: "WEBHOOK_SIGNATURE_VERIFICATION_FAILED",
	WEBHOOK_PAYLOAD_INVALID: "WEBHOOK_PAYLOAD_INVALID"
};
var ComposioFailedToGetSDKRealtimeCredentialsError = class extends ComposioError$1 {
	constructor(message = "Failed to get SDK realtime credentials", options = {}) {
		super(message, {
			...options,
			code: TriggerErrorCodes.TRIGGER_FAILED_TO_GET_SDK_REALTIME_CREDENTIALS,
			possibleFixes: options.possibleFixes || ["Please contact support."]
		});
		this.name = "ComposioFailedToGetSDKRealtimeCredentialsError";
	}
};
var ComposioFailedToCreatePusherClientError = class extends ComposioError$1 {
	constructor(message = "Failed to create Pusher client", options = {}) {
		super(message, {
			...options,
			code: TriggerErrorCodes.TRIGGER_FAILED_TO_CREATE_PUSHER_CLIENT,
			possibleFixes: options.possibleFixes || ["Please contact support."]
		});
		this.name = "ComposioFailedToCreatePusherClientError";
	}
};
var ComposioFailedToSubscribeToPusherChannelError = class extends ComposioError$1 {
	constructor(message = "Failed to subscribe to Pusher channel", options = {}) {
		super(message, {
			...options,
			code: TriggerErrorCodes.TRIGGER_FAILED_TO_SUBSCRIBE_TO_PUSHER_CHANNEL,
			possibleFixes: options.possibleFixes || ["Please contact support."]
		});
		this.name = "ComposioFailedToSubscribeToPusherChannelError";
	}
};
var ComposioFailedToUnsubscribeFromPusherChannelError = class extends ComposioError$1 {
	constructor(message = "Failed to unsubscribe from Pusher channel", options = {}) {
		super(message, {
			...options,
			code: TriggerErrorCodes.TRIGGER_FAILED_TO_UNSUBSCRIBE_FROM_PUSHER_CHANNEL,
			possibleFixes: options.possibleFixes || ["Please contact support."]
		});
		this.name = "ComposioFailedToUnsubscribeFromPusherChannelError";
	}
};
var ComposioTriggerTypeNotFoundError = class extends ComposioError$1 {
	constructor(message = "Trigger type not found", options = {}) {
		super(message, {
			...options,
			code: TriggerErrorCodes.TRIGGER_TYPE_NOT_FOUND,
			statusCode: 404,
			possibleFixes: options.possibleFixes || ["Please contact support."]
		});
		this.name = "ComposioTriggerTypeNotFoundError";
	}
};
var ComposioWebhookSignatureVerificationError = class extends ComposioError$1 {
	constructor(message = "Webhook signature verification failed", options = {}) {
		super(message, {
			...options,
			code: TriggerErrorCodes.WEBHOOK_SIGNATURE_VERIFICATION_FAILED,
			statusCode: 401,
			possibleFixes: options.possibleFixes || [
				"Verify that the webhook secret is correct.",
				"Ensure the raw request body is passed without modifications.",
				"Check that the signature header value is being passed correctly."
			]
		});
		this.name = "ComposioWebhookSignatureVerificationError";
	}
};
var ComposioWebhookPayloadError = class extends ComposioError$1 {
	constructor(message = "Invalid webhook payload", options = {}) {
		super(message, {
			...options,
			code: TriggerErrorCodes.WEBHOOK_PAYLOAD_INVALID,
			statusCode: 400,
			possibleFixes: options.possibleFixes || ["Ensure the webhook payload is valid JSON.", "Verify the payload structure matches the expected format."]
		});
		this.name = "ComposioWebhookPayloadError";
	}
};

//#endregion
//#region src/errors/RemoteFileErrors.ts
const RemoteFileErrorCodes = { DOWNLOAD_FAILED: "REMOTE_FILE_DOWNLOAD_FAILED" };
/**
* Error thrown when fetching a remote file from a tool router session mount fails.
* Includes HTTP status, URL, and file path context for debugging.
*/
var RemoteFileDownloadError = class extends ComposioError$1 {
	constructor(message = "Failed to download remote file", options = {}) {
		const { statusCode, statusText, downloadUrl, mountRelativePath, filename, meta: optionsMeta, ...rest } = options;
		const meta = {
			...optionsMeta,
			...statusCode !== void 0 && { statusCode },
			...statusText && { statusText },
			...downloadUrl && { downloadUrl },
			...mountRelativePath && { mountRelativePath },
			...filename && { filename }
		};
		super(message, {
			...rest,
			code: RemoteFileErrorCodes.DOWNLOAD_FAILED,
			statusCode,
			meta: Object.keys(meta).length > 0 ? meta : void 0,
			possibleFixes: options.possibleFixes ?? [
				"Verify the download URL has not expired (check expiresAt)",
				"Ensure the file exists at the specified mount path",
				"Retry the operation; presigned URLs may have transient failures"
			]
		});
		this.name = "RemoteFileDownloadError";
	}
};

//#endregion
export { ToolErrorCodes as A, ComposioSharedAccessDeniedError as B, ComposioInvalidExecuteFunctionError as C, ComposioToolExecutionError as D, ComposioProviderNotDefinedError as E, ComposioAclOnlyForSharedError as F, ConnectedAccountErrorCodes as H, ComposioConnectedAccountNotFoundError as I, ComposioFailedToCreateConnectedAccountLink as L, ConnectionRequestErrorCodes as M, ConnectionRequestFailedError as N, ComposioToolNotFoundError as O, ConnectionRequestTimeoutError as P, ComposioLegacyConnectedAccountsEndpointRetiredError as R, ComposioGlobalExecuteToolFnNotSetError as S, ComposioInvalidToolArgumentsError as T, AuthConfigErrorCodes as U, ComposioSharedConnectionNotAccessibleError as V, ComposioAuthConfigNotFoundError as W, ValidationError as _, ComposioFailedToSubscribeToPusherChannelError as a, ComposioToolkitNotFoundError as b, ComposioWebhookPayloadError as c, ComposioNoAPIKeyError as d, ComposioRequestCancelledError as f, JsonSchemaToZodError as g, JsonSchemaRefResolutionError as h, ComposioFailedToGetSDKRealtimeCredentialsError as i, handleToolExecutionError as j, ComposioToolVersionRequiredError as k, ComposioWebhookSignatureVerificationError as l, isRequestAbortError as m, RemoteFileErrorCodes as n, ComposioFailedToUnsubscribeFromPusherChannelError as o, SDKErrorCodes as p, ComposioFailedToCreatePusherClientError as r, ComposioTriggerTypeNotFoundError as s, RemoteFileDownloadError as t, TriggerErrorCodes as u, ValidationErrorCodes as v, ComposioInvalidModifierError as w, ToolkitErrorCodes as x, ComposioToolkitFetchError as y, ComposioMultipleConnectedAccountsError as z };
