// @composio/core/dist/fileUtils.node-iPFi2o4i.mjs
import { d as COMPOSIO_DIR, g as TEMP_FILES_DIRECTORY_NAME, o as ComposioFileUploadPathNotAllowedError, r as ComposioFileNotFoundError, t as ComposioBlockedInternalUrlError, u as logger } from "./SsrfErrors-BQRozkeU.mjs";
import { a as base64ToUint8Array, n as assertSafeFileUploadPath, o as uint8ArrayToBase64, s as getRandomShortId } from "./sensitiveFileUploadPaths-B1OLsd8F.mjs";
import { platform } from "#platform";
import fs from "node:fs";
import * as path from "node:path";
import crypto from "node:crypto";
import { lookup } from "node:dns/promises";
import { isIP } from "node:net";

//#region src/utils/uploadDirAllowlist.node.ts
/**
* Allowlist for automatic file upload during tool execution.
*
* This is only consulted when `dangerouslyAllowAutoUploadDownloadFiles` is enabled.
* Manual `composio.files.upload()` calls are not subject to the allowlist.
*
* - A local path is accepted iff its real (symlink-resolved, absolute) path is
*   inside one of the allowed directories on a path-component boundary.
*   `/tmp/foo` allows `/tmp/foo/bar` but NOT `/tmp/foo-bar`.
* - URLs (`http(s)://…`) and `File`/`Blob` values never hit this check.
* - User-provided `fileUploadDirs` fully replaces the default `[~/.composio/temp]`.
* - On Windows, path comparisons are case-insensitive.
*/
function normalizeForCompare(p) {
	return process.platform === "win32" ? p.toLowerCase() : p;
}
/**
* True if `child` is equal to `parent` or nested inside it on a component
* boundary. Assumes both paths are absolute and normalized.
*/
function isInsideDir(child, parent) {
	const a = normalizeForCompare(child);
	const b = normalizeForCompare(parent);
	if (a === b) return true;
	const sep = path.sep;
	const bWithSep = b.endsWith(sep) ? b : b + sep;
	return a.startsWith(bWithSep);
}
/**
* Returns the real (symlink-resolved) absolute path when the target exists,
* else `null`. Mirrors the semantics of the sensitive-path util.
*/
function tryRealpath(absPath) {
	try {
		if (fs.existsSync(absPath)) return fs.realpathSync(absPath);
	} catch {}
	return null;
}
function formatAllowlist(dirs) {
	if (dirs.length === 0) return "  (none configured — all local paths are blocked)";
	return dirs.map((d) => `  - ${d}`).join("\n");
}
function buildHelpFooter(allowlist) {
	return [
		"",
		"Allowed upload directories (fileUploadDirs):",
		formatAllowlist(allowlist),
		"",
		"Fix one of the following:",
		"",
		"  1. Recommended — move the file under an allowed directory, or add your",
		"     directory to the allowlist:",
		"       new Composio({",
		"         fileUploadDirs: ['/abs/path/to/uploads', '~/.composio/temp'],",
		"       })",
		"     Note: user-provided `fileUploadDirs` REPLACES the default",
		"     `~/.composio/temp`. Include it explicitly if you want staged uploads",
		"     to keep working.",
		"",
		"  2. Pass the file by URL or as a File/Blob object instead of a filesystem",
		"     path. URLs and File objects are never path-checked.",
		"",
		"  3. (Dangerous) Bypass the allowlist entirely by opting into automatic",
		"     file upload/download from any readable path:",
		"       new Composio({ dangerouslyAllowAutoUploadDownloadFiles: true })",
		"     WARNING: This lets a prompt-injected or misbehaving tool read ANY",
		"     file the process can access. Only enable it in trusted, sandboxed",
		"     environments. The built-in sensitive-path denylist (.ssh, .aws,",
		"     .env, private SSH keys, etc.) still applies unless you also set",
		"     `sensitiveFileUploadProtection: false`."
	].join("\n");
}
/**
* @throws {ComposioFileNotFoundError} when `filePath` does not exist on disk.
* @throws {ComposioFileUploadPathNotAllowedError} when the resolved path is
*         outside every entry of `allowlist`.
*/
function assertPathInsideUploadDirs(filePath, allowlist) {
	const attempted = filePath;
	const abs = path.resolve(filePath);
	const real = tryRealpath(abs);
	if (!real) {
		const cwd = process.cwd();
		const parent = path.dirname(abs);
		const parentExists = (() => {
			try {
				return fs.existsSync(parent);
			} catch {
				return false;
			}
		})();
		throw new ComposioFileNotFoundError([
			`Refusing to auto-upload "${attempted}": the file does not exist on disk.`,
			"",
			`Path attempted:   ${attempted}`,
			`Resolved to:      ${abs}`,
			`Process cwd:      ${cwd}`,
			`Parent exists:    ${parentExists ? "yes (" + parent + ")" : "no (" + parent + ")"}`,
			"",
			"Common causes:",
			"  - Typo in the filename passed to the tool.",
			"  - Relative path resolved against the wrong working directory",
			"    (relative paths use process.cwd() at the moment of upload).",
			"  - File was deleted between the tool being called and the upload starting.",
			"",
			buildHelpFooter(allowlist)
		].join("\n"), { meta: {
			attempted,
			resolved: abs,
			cwd,
			allowlist
		} });
	}
	if (allowlist.length === 0) throw new ComposioFileUploadPathNotAllowedError([
		`Refusing to auto-upload "${attempted}": no upload directories are configured.`,
		"",
		`Path attempted:   ${attempted}`,
		`Resolved to:      ${real}`,
		"",
		"Automatic file upload during tool execution is locked down by default",
		"to prevent a prompt-injected tool from exfiltrating server files",
		"(source code, .env, SSH keys, etc.).",
		buildHelpFooter(allowlist)
	].join("\n"), { meta: {
		attempted,
		resolved: real,
		allowlist
	} });
	for (const dir of allowlist) if (isInsideDir(real, tryRealpath(dir) ?? path.resolve(dir))) return;
	throw new ComposioFileUploadPathNotAllowedError([
		`Refusing to auto-upload "${attempted}": resolved path is not inside any`,
		"directory in the configured `fileUploadDirs` allowlist.",
		"",
		`Path attempted:   ${attempted}`,
		`Resolved to:      ${real}`,
		buildHelpFooter(allowlist)
	].join("\n"), { meta: {
		attempted,
		resolved: real,
		allowlist
	} });
}

//#endregion
//#region src/utils/ssrfGuard.node.ts
/**
* SSRF guard for user-supplied URL file inputs.
*
* `composio.files.upload(url)` and automatic file upload during tool execution
* fetch arbitrary user-provided URLs. Without a guard, a caller (or a tool
* argument produced by an LLM) can point the SDK at internal infrastructure —
* loopback, RFC1918 ranges, link-local cloud-metadata endpoints
* (`169.254.169.254`), or a public URL that 3xx-redirects into internal space —
* turning the SDK into a server-side request forgery probe.
*
* This module validates the *resolved* address (not just the hostname string,
* which defeats decimal/octal/hex IP obfuscation) before every fetch, and
* follows redirects manually so each hop is re-validated.
*
* Known residual: a DNS name could resolve to a public address here and rebind
* to an internal one before the underlying `fetch` connects (a TOCTOU / DNS
* rebinding window). Closing that fully requires pinning the validated IP at
* connect time (a custom dispatcher), which is out of scope for this guard.
*/
const MAX_REDIRECTS = 5;
const ipv4ToLong = (ip) => {
	const [a, b, c, d] = ip.split(".").map(Number);
	return (a << 24 >>> 0 | b << 16 | c << 8 | d) >>> 0;
};
const IPV4_BLOCKED_CIDRS = [
	["0.0.0.0", 8],
	["10.0.0.0", 8],
	["100.64.0.0", 10],
	["127.0.0.0", 8],
	["169.254.0.0", 16],
	["172.16.0.0", 12],
	["192.0.0.0", 24],
	["192.0.2.0", 24],
	["192.168.0.0", 16],
	["198.18.0.0", 15],
	["198.51.100.0", 24],
	["203.0.113.0", 24],
	["224.0.0.0", 4],
	["240.0.0.0", 4]
];
const isBlockedIpv4Long = (long) => IPV4_BLOCKED_CIDRS.some(([base, bits]) => {
	const mask = bits === 0 ? 0 : 4294967295 << 32 - bits >>> 0;
	return (long & mask) === (ipv4ToLong(base) & mask);
});
/**
* Expand an IPv6 string into its 8 16-bit groups, converting any trailing
* dotted-quad (e.g. `::ffff:1.2.3.4`) into hex groups first. Returns `null` when
* the input cannot be parsed.
*/
const expandIpv6 = (ip) => {
	let addr = ip;
	const v4Match = addr.match(/^(.*:)(\d+\.\d+\.\d+\.\d+)$/);
	if (v4Match) {
		const long = ipv4ToLong(v4Match[2]);
		addr = `${v4Match[1]}${(long >>> 16 & 65535).toString(16)}:${(long & 65535).toString(16)}`;
	}
	const halves = addr.split("::");
	if (halves.length > 2) return null;
	const head = halves[0] ? halves[0].split(":") : [];
	const tail = halves.length === 2 ? halves[1] ? halves[1].split(":") : [] : [];
	let groups;
	if (halves.length === 2) {
		const missing = 8 - head.length - tail.length;
		if (missing < 0) return null;
		groups = [
			...head,
			...Array(missing).fill("0"),
			...tail
		];
	} else groups = head;
	if (groups.length !== 8) return null;
	return groups.map((g) => parseInt(g || "0", 16));
};
const isBlockedIpv6 = (ip) => {
	const h = expandIpv6(ip);
	if (!h) return true;
	if (h.every((x) => x === 0)) return true;
	if (h.slice(0, 7).every((x) => x === 0) && h[7] === 1) return true;
	const isMappedOrCompat = h[0] === 0 && h[1] === 0 && h[2] === 0 && h[3] === 0 && h[4] === 0 && (h[5] === 65535 || h[5] === 0);
	const isNat64 = h[0] === 100 && h[1] === 65435 && h[2] === 0 && h[3] === 0 && h[4] === 0 && h[5] === 0;
	if (isMappedOrCompat || isNat64) return isBlockedIpv4Long((h[6] << 16 >>> 0 | h[7]) >>> 0);
	if ((h[0] & 65024) === 64512) return true;
	if ((h[0] & 65472) === 65152) return true;
	if ((h[0] & 65280) === 65280) return true;
	return false;
};
/**
* True when `ip` is a private, loopback, link-local, or otherwise
* non-publicly-routable address. Fails closed (returns `true`) for anything that
* is not a valid IPv4/IPv6 literal.
*/
const isBlockedIp = (ip) => {
	const family = isIP(ip);
	if (family === 4) return isBlockedIpv4Long(ipv4ToLong(ip));
	if (family === 6) return isBlockedIpv6(ip);
	return true;
};
/**
* Validate a single URL: it must be http(s), its host must resolve, and every
* resolved address must be publicly routable. Throws
* {@link ComposioBlockedInternalUrlError} otherwise.
*/
const assertSafeFetchTarget = async (rawUrl) => {
	let url;
	try {
		url = new URL(rawUrl);
	} catch {
		throw new ComposioBlockedInternalUrlError("Refusing to fetch a malformed URL", { url: rawUrl });
	}
	if (url.protocol !== "http:" && url.protocol !== "https:") throw new ComposioBlockedInternalUrlError(`Refusing to fetch a non-http(s) URL (scheme "${url.protocol}")`, { url: rawUrl });
	const host = url.hostname.replace(/^\[|\]$/g, "");
	let resolved;
	try {
		resolved = await lookup(host, {
			all: true,
			verbatim: true
		});
	} catch {
		throw new ComposioBlockedInternalUrlError(`Could not resolve host "${host}"`, { url: rawUrl });
	}
	for (const { address } of resolved) if (isBlockedIp(address)) throw new ComposioBlockedInternalUrlError(`Refusing to fetch "${host}" — it resolves to a private, loopback, or link-local address`, {
		url: rawUrl,
		resolvedIp: address
	});
};
/**
* Drop-in replacement for `fetch` that blocks SSRF. Validates the target before
* connecting and re-validates every redirect hop (redirects are followed
* manually up to {@link MAX_REDIRECTS}). Non-redirect responses are returned
* unchanged.
*/
const ssrfSafeFetch = async (rawUrl, init = {}, maxRedirects = MAX_REDIRECTS) => {
	let currentUrl = rawUrl;
	for (let hop = 0; hop <= maxRedirects; hop++) {
		await assertSafeFetchTarget(currentUrl);
		const response = await fetch(currentUrl, {
			...init,
			redirect: "manual"
		});
		if (!(response.status >= 300 && response.status < 400 && response.headers.has("location"))) return response;
		currentUrl = new URL(response.headers.get("location"), currentUrl).toString();
	}
	throw new ComposioBlockedInternalUrlError(`Refusing to fetch: too many redirects (max ${maxRedirects})`, { url: rawUrl });
};

//#endregion
//#region src/utils/fileUtils.node.ts
/**
* Checks whether `value` is an absolute http(s) URL rather than a local path.
*
* A bare `startsWith('http')` check also matches local paths like
* `http_export/report.csv` or `httpdocs/.env`, which routes them into the URL
* fetch branch and skips the allowlist/denylist checks in
* {@link getFileDataAfterUploadingToS3}. `new URL()` requires a real scheme
* (`http:` or `https:`) followed by `//`, so relative-looking paths correctly
* fail to parse and are treated as local.
*/
const isHttpUrl = (value) => {
	try {
		const url = new URL(value);
		return url.protocol === "http:" || url.protocol === "https:";
	} catch {
		return false;
	}
};
const getExtensionFromMimeType = (mimeType) => {
	const mimeToExt = {
		"text/plain": "txt",
		"text/html": "html",
		"text/css": "css",
		"text/javascript": "js",
		"application/json": "json",
		"application/xml": "xml",
		"application/pdf": "pdf",
		"application/zip": "zip",
		"application/x-zip-compressed": "zip",
		"application/gzip": "gz",
		"application/x-tar": "tar",
		"application/msword": "doc",
		"application/vnd.openxmlformats-officedocument.wordprocessingml.document": "docx",
		"application/vnd.ms-excel": "xls",
		"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": "xlsx",
		"application/vnd.ms-powerpoint": "ppt",
		"application/vnd.openxmlformats-officedocument.presentationml.presentation": "pptx",
		"image/jpeg": "jpg",
		"image/jpg": "jpg",
		"image/png": "png",
		"image/gif": "gif",
		"image/svg+xml": "svg",
		"image/webp": "webp",
		"image/bmp": "bmp",
		"image/tiff": "tiff",
		"audio/mpeg": "mp3",
		"audio/wav": "wav",
		"audio/ogg": "ogg",
		"video/mp4": "mp4",
		"video/mpeg": "mpeg",
		"video/quicktime": "mov",
		"video/x-msvideo": "avi",
		"video/webm": "webm"
	};
	const cleanMimeType = mimeType.split(";")[0].toLowerCase().trim();
	if (mimeToExt[cleanMimeType]) return mimeToExt[cleanMimeType];
	const parts = cleanMimeType.split("/");
	if (parts.length === 2) {
		const cleanSubtype = parts[1].toLowerCase();
		if (cleanSubtype.includes("+")) {
			const plusParts = cleanSubtype.split("+");
			const prefix = plusParts[0];
			const suffix = plusParts[plusParts.length - 1];
			if ([
				"svg",
				"atom",
				"rss"
			].includes(prefix)) return prefix;
			if ([
				"json",
				"xml",
				"yaml",
				"zip",
				"gzip"
			].includes(suffix)) return suffix;
			return suffix;
		}
		return cleanSubtype || "txt";
	}
	return "txt";
};
const generateTimestampedFilename = (extension, prefix) => {
	return `${prefix || "file_ts"}${Date.now()}${getRandomShortId()}.${extension}`;
};
const readFileContent = async (filePath) => {
	try {
		if (!platform.supportsFileSystem) throw new Error("File system operations are not supported in this runtime environment");
		const content = platform.readFileSync(filePath);
		return {
			fileName: generateTimestampedFilename(filePath.split(".").pop() || "txt"),
			content: content instanceof Uint8Array ? uint8ArrayToBase64(content) : uint8ArrayToBase64(new TextEncoder().encode(content)),
			mimeType: "application/octet-stream"
		};
	} catch (error) {
		throw new Error(`Error reading file at ${filePath}: ${error}`);
	}
};
const readFileContentFromURL = async (path, signal) => {
	const response = await ssrfSafeFetch(path, { signal });
	if (!response.ok) throw new Error(`Failed to fetch file: ${response.statusText}`);
	const arrayBuffer = await response.arrayBuffer();
	const content = new Uint8Array(arrayBuffer);
	const mimeType = response.headers.get("content-type") || "application/octet-stream";
	const pathname = new URL(path).pathname;
	let fileName = platform.basename(pathname);
	if (!fileName || fileName === "/") {
		const extension = getExtensionFromMimeType(mimeType);
		fileName = generateTimestampedFilename(extension);
	} else if (!fileName.includes(".")) {
		const extension = getExtensionFromMimeType(mimeType);
		fileName = generateTimestampedFilename(extension);
	}
	return {
		content: uint8ArrayToBase64(content),
		mimeType,
		fileName
	};
};
const uploadFileToS3 = async (fileName, content, toolSlug, toolkitSlug, mimeType, client, signal) => {
	const contentBytes = base64ToUint8Array(content);
	const { key, new_presigned_url: signedURL } = await client.files.createPresignedURL({
		filename: fileName,
		mimetype: mimeType,
		md5: crypto.createHash("md5").update(contentBytes).digest("hex"),
		tool_slug: toolSlug,
		toolkit_slug: toolkitSlug
	}, { signal });
	logger.debug(`Uploading ${key} file to S3: ${key}`);
	const uploadBuffer = new Uint8Array(contentBytes.byteLength);
	uploadBuffer.set(contentBytes);
	const uploadResponse = await fetch(signedURL, {
		method: "PUT",
		body: uploadBuffer,
		signal,
		headers: { "Content-Type": mimeType }
	});
	if (!uploadResponse.ok) throw new Error(`Failed to upload file to S3: ${uploadResponse.statusText}`);
	return key;
};
const readFile = async (file, signal) => {
	if (file instanceof File) {
		const content = await file.arrayBuffer();
		return {
			fileName: file.name,
			content: uint8ArrayToBase64(new Uint8Array(content)),
			mimeType: file.type
		};
	} else if (typeof file === "string") if (isHttpUrl(file)) return await readFileContentFromURL(file, signal);
	else return await readFileContent(file);
	throw new Error("Invalid file type");
};
const getFileDataAfterUploadingToS3 = async (file, { toolSlug, toolkitSlug, client, sensitiveFileUploadProtection, fileUploadPathDenySegments, fileUploadAllowlist, signal }) => {
	if (!file) throw new Error("Either path or blob must be provided");
	const isLocalPath = typeof file === "string" && !isHttpUrl(file);
	if (isLocalPath && fileUploadAllowlist !== void 0) assertPathInsideUploadDirs(file, fileUploadAllowlist);
	if (sensitiveFileUploadProtection !== false && isLocalPath) assertSafeFileUploadPath(file, { additionalDenySegments: fileUploadPathDenySegments });
	const fileData = await readFile(file, signal);
	logger.debug(`Uploading file to S3...`);
	const s3key = await uploadFileToS3(platform.basename(fileData.fileName), fileData.content, toolSlug, toolkitSlug, fileData.mimeType, client, signal);
	logger.debug(`Done! File uploaded to S3: ${s3key}`, JSON.stringify(fileData, null, 2));
	return {
		name: fileData.fileName,
		mimetype: fileData.mimeType,
		s3key
	};
};
const downloadFileFromS3 = async ({ toolSlug, s3Url, mimeType, fileDownloadDir, signal }) => {
	const response = await fetch(s3Url, { signal });
	if (!response.ok) throw new Error(`Failed to download file: ${response.statusText}`);
	const data = await response.arrayBuffer();
	const extension = getExtensionFromMimeType(mimeType);
	const fileName = generateTimestampedFilename(extension, `${toolSlug}_`);
	return {
		name: fileName,
		mimeType,
		s3Url,
		/**
		* @todo: fix in follow-up PR.
		*/
		filePath: saveFile(fileName, new Uint8Array(data), {
			isTempFile: fileDownloadDir === void 0,
			outputDir: fileDownloadDir
		})
	};
};
/**
* Gets the Composio directory.
* @param createDirIfNotExists - Whether to create the directory if it doesn't exist.
* @returns The path to the Composio directory.
*/
const getComposioDir = (createDirIfNotExists = false) => {
	try {
		const homeDir = platform.homedir();
		if (!homeDir) return null;
		const composioDir = platform.joinPath(homeDir, COMPOSIO_DIR);
		if (createDirIfNotExists && platform.supportsFileSystem && !platform.existsSync(composioDir)) platform.mkdirSync(composioDir);
		return composioDir;
	} catch (_error) {
		return null;
	}
};
/**
* Gets the Composio temporary files directory.
* @param createDirIfNotExists - Whether to create the directory if it doesn't exist.
* @returns The path to the Composio temporary files directory.
*/
const getComposioTempFilesDir = (createDirIfNotExists = false) => {
	try {
		const homeDir = platform.homedir();
		if (!homeDir) return null;
		const composioFilesDir = platform.joinPath(homeDir, COMPOSIO_DIR, TEMP_FILES_DIRECTORY_NAME);
		if (createDirIfNotExists && platform.supportsFileSystem && !platform.existsSync(composioFilesDir)) platform.mkdirSync(composioFilesDir);
		return composioFilesDir;
	} catch (_error) {
		return null;
	}
};
/**
* Saves a file to a local directory.
*
* Directory precedence (highest first):
*   1. `options.outputDir` (explicit, absolute or relative; created if missing)
*   2. `options.isTempFile === true` → `<home>/.composio/files`
*   3. fallback → `<home>/.composio`
*
* @returns The absolute path to the saved file, or `null` if the runtime has
*          no filesystem (e.g. Cloudflare Workers) or the save failed.
*/
const saveFile = (file, content, options = {}) => {
	const opts = typeof options === "boolean" ? { isTempFile: options } : options;
	try {
		if (!platform.supportsFileSystem) {
			logger.debug("File system operations are not supported in this runtime environment");
			return null;
		}
		let composioFilesDir;
		if (opts.outputDir) {
			const absDir = platform.joinPath(opts.outputDir);
			try {
				if (!platform.existsSync(absDir)) platform.mkdirSync(absDir);
				composioFilesDir = absDir;
			} catch (err) {
				logger.warn(`fileDownloadDir "${absDir}" is not writable, falling back to the default (~/.composio/files). Error: ${err}`);
				composioFilesDir = getComposioTempFilesDir(true);
			}
		} else composioFilesDir = opts.isTempFile ? getComposioTempFilesDir(true) : getComposioDir(true);
		if (!composioFilesDir) return null;
		const filePath = platform.joinPath(composioFilesDir, platform.basename(file));
		logger.info(`Saving file to: ${filePath}`);
		if (content instanceof Uint8Array) platform.writeFileSync(filePath, content);
		else platform.writeFileSync(filePath, content, "utf8");
		return filePath;
	} catch (_error) {
		logger.debug(`Error saving file: ${_error}`);
		return null;
	}
};

//#endregion
export { getFileDataAfterUploadingToS3 as n, isHttpUrl as r, downloadFileFromS3 as t };
