// @composio/core/dist/sensitiveFileUploadPaths-B1OLsd8F.mjs
import { s as ComposioSensitiveFilePathBlockedError } from "./SsrfErrors-BQRozkeU.mjs";
import { platform } from "#platform";

//#region src/utils/uuid.ts
function getRandomUUID() {
	return globalThis.crypto.randomUUID();
}
function getRandomShortId() {
	return getRandomUUID().slice(0, 8).replace(/-/g, "");
}

//#endregion
//#region src/utils/buffer.ts
/**
* Convert ArrayBuffer to base64.
*/
function arrayBufferToBase64(buffer) {
	const bytes = new Uint8Array(buffer);
	let binary = "";
	for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
	return btoa(binary);
}
const base64ToUint8Array = (base64) => {
	const binary = atob(base64);
	const bytes = new Uint8Array(binary.length);
	for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
	return bytes;
};
const uint8ArrayToBase64 = (bytes) => {
	let binary = "";
	for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
	return btoa(binary);
};

//#endregion
//#region src/utils/sensitiveFileUploadPaths.ts
/**
* Blocks accidental upload of local files from well-known secret/credential locations
* during auto file upload (and {@link getFileDataAfterUploadingToS3}).
*
* This is the single canonical guard shared by the core SDK and downstream
* packages (e.g. `@composio/cli`). Filesystem access goes through the `#platform`
* abstraction so this module stays free of static `node:*` imports and is safe to
* re-export from the package root (edge/workerd builds included).
*/
/**
* Path segments (a single path component) that indicate a sensitive directory when
* they appear anywhere in a resolved local path. Compared case-insensitively on Windows.
*/
const BUILTIN_FILE_UPLOAD_PATH_DENY_SEGMENTS = [
	".ssh",
	".aws",
	".azure",
	".gnupg",
	".kube",
	".docker",
	".claude",
	".password-store",
	"keychains"
];
const SECRET_LIKE_BASENAME = /^(\.env(\.|$)|\.netrc$|\.pgpass$)/i;
/** Default SSH private key basenames (public keys like id_rsa.pub are allowed). */
const DEFAULT_PRIVATE_KEY_BASENAME = /^id_(rsa|ed25519|ecdsa|dsa|ecdsa_sk)(\.old)?$/i;
const isWindows = typeof process !== "undefined" && process.platform === "win32";
/**
* Returns normalized path segments, resolving symlinks when the path exists.
*/
function normalizePathSegments(filePath) {
	const absolute = platform.resolvePath(filePath);
	let resolved = absolute;
	try {
		if (platform.existsSync(absolute)) resolved = platform.realpathSync(absolute);
	} catch {}
	return resolved.split(/[/\\]+/).filter(Boolean);
}
/**
* True if the path is under a built-in or extra deny segment, or the basename
* looks like an env / netrc / default SSH private key name.
*/
function isBlockedSensitiveFileUploadPath(filePath, additionalDenySegments) {
	return getSensitiveFileUploadPathBlockReason(filePath, additionalDenySegments) != null;
}
function getSensitiveFileUploadPathBlockReason(filePath, additionalDenySegments) {
	const segments = normalizePathSegments(filePath);
	const deny = new Set([...BUILTIN_FILE_UPLOAD_PATH_DENY_SEGMENTS, ...(additionalDenySegments ?? []).map((s) => s.trim()).filter(Boolean)].map((s) => isWindows ? s.toLowerCase() : s));
	const segmentsForMatch = isWindows ? segments.map((s) => s.toLowerCase()) : segments;
	for (let i = 0; i < segments.length; i++) if (deny.has(segmentsForMatch[i])) return `path segment "${segments[i]}" is in the sensitive file upload denylist`;
	const basename = segments.length > 0 ? segments[segments.length - 1] : "";
	if (basename) {
		if (SECRET_LIKE_BASENAME.test(basename) || DEFAULT_PRIVATE_KEY_BASENAME.test(basename)) return `file name "${basename}" looks like a credential, env, or private key file`;
		if (basename.toLowerCase() === "credentials") return "file name \"credentials\" is often used for cloud/API credential stores";
	}
	return null;
}
/**
* Default remediation hint appended to the block message. Assumes an SDK caller
* that exposes `sensitiveFileUploadProtection`. Callers without such an opt-out
* (e.g. `@composio/cli`) should pass their own `remediation` so the message does
* not advertise an option the caller cannot honor.
*/
const DEFAULT_REMEDIATION = "To upload from this path anyway, set sensitiveFileUploadProtection: false on Composio (not recommended) or use a copy outside sensitive locations.";
/**
* @throws {ComposioSensitiveFilePathBlockedError} if the path is not allowed
*/
function assertSafeFileUploadPath(filePath, options) {
	const reason = getSensitiveFileUploadPathBlockReason(filePath, options?.additionalDenySegments);
	if (reason) throw new ComposioSensitiveFilePathBlockedError(`Refusing to upload: ${reason}. ${options?.remediation ?? DEFAULT_REMEDIATION}`, { meta: {
		filePath,
		reason
	} });
}

//#endregion
export { base64ToUint8Array as a, getRandomUUID as c, arrayBufferToBase64 as i, assertSafeFileUploadPath as n, uint8ArrayToBase64 as o, isBlockedSensitiveFileUploadPath as r, getRandomShortId as s, BUILTIN_FILE_UPLOAD_PATH_DENY_SEGMENTS as t };
