// @composio/core/dist/Telemetry-llfvticE.mjs
import { l as ComposioError$1, u as logger, y as getEnvVariable } from "./SsrfErrors-BQRozkeU.mjs";
import { c as getRandomUUID } from "./sensitiveFileUploadPaths-B1OLsd8F.mjs";
import { ComposioError } from "@composio/client";

//#region src/types/telemetry.types.ts
const TELEMETRY_EVENTS = {
	SDK_INITIALIZED: "SDK_INITIALIZED",
	SDK_METHOD_INVOKED: "SDK_METHOD_INVOKED",
	SDK_METHOD_ERROR: "SDK_METHOD_ERROR",
	CLI_INVOKED: "CLI_INVOKED"
};

//#endregion
//#region src/telemetry/BatchProcessor.ts
var BatchProcessor = class {
	batch = [];
	time;
	batchSize;
	processBatchCallback;
	timer = null;
	pendingBatches = /* @__PURE__ */ new Set();
	constructor(time = 2e3, batchSize = 100, processBatchCallback) {
		this.batch = [];
		this.time = time;
		this.batchSize = batchSize;
		this.processBatchCallback = processBatchCallback;
	}
	pushItem(item) {
		this.batch.push(item);
		if (this.batch.length >= this.batchSize) this.processBatch();
		else if (!this.timer) this.timer = setTimeout(() => this.processBatch(), this.time);
	}
	processBatch() {
		if (this.batch.length > 0) {
			const batchToProcess = this.batch;
			this.batch = [];
			const pending = new Promise((resolve) => {
				const run = () => {
					Promise.resolve().then(() => this.processBatchCallback(batchToProcess)).catch(() => {}).finally(resolve);
				};
				if (typeof queueMicrotask === "function") queueMicrotask(run);
				else setTimeout(run, 0);
			});
			this.pendingBatches.add(pending);
			pending.finally(() => {
				this.pendingBatches.delete(pending);
			});
		}
		if (this.timer) {
			clearTimeout(this.timer);
			this.timer = null;
		}
	}
	/**
	* Flush any pending batches and wait for all of them to complete.
	* Useful for ensuring telemetry is sent before process exit.
	*/
	async flush() {
		this.processBatch();
		if (this.pendingBatches.size > 0) await Promise.all(this.pendingBatches);
	}
};

//#endregion
//#region src/services/telemetry/TelemetryService.ts
const TELEMETRY_URL = "https://telemetry.composio.dev/v1";
var TelemetryService = class {
	/**
	* Sends a metric to the Telemetry API.
	* @param payload - The payload to send to the Telemetry API.
	* @returns The response from the Telemetry API.
	*/
	static async sendMetric(payload) {
		try {
			return await fetch(`${TELEMETRY_URL}/metrics/invocations`, {
				method: "POST",
				headers: { "Content-Type": "application/json" },
				body: JSON.stringify(payload)
			});
		} catch (error) {
			logger.debug("Error sending metric telemetry", error);
		}
	}
	/**
	* Sends an error log to the Telemetry API.
	* @param payload - The payload to send to the Telemetry API.
	* @returns The response from the Telemetry API.
	*/
	static async sendErrorLog(payload) {
		try {
			return await fetch(`${TELEMETRY_URL}/errors`, {
				method: "POST",
				headers: { "Content-Type": "application/json" },
				body: JSON.stringify(payload)
			});
		} catch (error) {
			logger.debug("Error sending error telemetry", error);
		}
	}
};

//#endregion
//#region src/telemetry/redact.ts
/**
* Best-effort redaction of secrets from free-form error text before it is sent
* to the telemetry endpoint.
*
* Error `message` and `stack` fields routinely interpolate URLs (with
* query-string tokens / presigned signatures), `Authorization` headers, API
* keys, and connected-account identifiers. None of that should leave the
* process. Structured telemetry fields never carry raw secrets, so only the two
* free-form strings are passed through here.
*
* This is defence-in-depth, not a proof: it targets the shapes we have seen
* leak. When in doubt it over-redacts rather than under-redacts.
*/
const REDACTED = "[REDACTED]";
const REDACTION_RULES = [
	[/(\bhttps?:\/\/[^\s?#'"]+)\?[^\s'"]*/gi, `$1?${REDACTED}`],
	[/\b(bearer|basic)\s+[A-Za-z0-9._~+/=-]+/gi, `$1 ${REDACTED}`],
	[/\b(authorization|api[-_]?key|apikey|x-api-key|access[-_]?token|refresh[-_]?token|client[-_]?secret|secret|password|passwd|pwd)\b(\s*[:=]+\s*)(["']?)([^\s"',}&]+)\3/gi, `$1$2$3${REDACTED}$3`]
];
/**
* Redact common secret shapes from a free-form string. Returns the input
* unchanged when it is empty or `undefined`.
*/
const redactSensitiveText = (input) => {
	if (!input) return input;
	let output = input;
	for (const [pattern, replacement] of REDACTION_RULES) output = output.replace(pattern, replacement);
	return output;
};

//#endregion
//#region src/telemetry/Telemetry.ts
/**
* The Telemetry class is used to log the telemetry for any given instance which extends InstrumentedInstance.
*
* This class is used to instrument the telemetry for the given instance and send the telemetry to the server.
* This class is also used to create a global error handler for the given instance.
*
* @example
*
* const telemetry = new Telemetry({...});
* const composio = new Composio({...})
*
* telemetry.instrument(composio);
* telemetry.instrument(composio.tools);
* telemetry.instrument(composio.toolkits);
* telemetry.instrument(composio.triggers);
*
*/
var TelemetryTransport = class {
	telemetryMetadata;
	isTelemetryDisabled = true;
	telemetrySource;
	telemetrySourceName = "typescript-sdk";
	telemetryServiceName = "sdk";
	telemetryLanguage = "typescript";
	exitHandlersRegistered = false;
	pendingImmediateTelemetry = /* @__PURE__ */ new Set();
	batchProcessor = new BatchProcessor(200, 10, async (data) => {
		logger.debug("Sending batch of telemetry metrics", data);
		await TelemetryService.sendMetric(data);
	});
	setup(metadata) {
		this.telemetryMetadata = metadata;
		this.isTelemetryDisabled = false;
		this.telemetrySource = {
			host: this.telemetryMetadata?.host ?? this.telemetrySourceName,
			service: this.telemetryServiceName,
			language: this.telemetryLanguage,
			version: this.telemetryMetadata?.version,
			platform: this.telemetryMetadata?.isBrowser ? "browser" : "node",
			environment: getEnvVariable("NODE_ENV", "production")
		};
		this.registerExitHandlers();
		this.deferTelemetry(() => this.sendMetric([{
			functionName: TELEMETRY_EVENTS.SDK_INITIALIZED,
			durationMs: 0,
			timestamp: Date.now() / 1e3,
			props: {},
			source: this.telemetrySource,
			metadata: { provider: this.telemetryMetadata?.provider ?? "openai" },
			error: void 0
		}]));
	}
	/**
	* Instrument the telemetry for the given instance.
	*
	* You can pass the instance and the file name of the instance to instrument the telemetry.
	* This will instrument all the methods of the instance and log the telemetry for each method call.
	* @param instance - any instance that extends InstrumentedInstance
	* @param fileName - the file name of the instance
	* @returns
	*/
	instrument(instance, fileName) {
		const proto = Object.getPrototypeOf(instance);
		const methodNames = Object.getOwnPropertyNames(proto).filter((key) => {
			const descriptor = Object.getOwnPropertyDescriptor(proto, key);
			return key !== "constructor" && descriptor && typeof descriptor.value === "function" && descriptor.value.constructor.name === "AsyncFunction";
		});
		const instrumentedClassName = fileName ?? instance.constructor?.name ?? "unknown";
		for (const name of methodNames) {
			const originalMethod = instance[name];
			instance[name] = async (...args) => {
				const telemetryEnabled = this.shouldSendTelemetry();
				const startTime = telemetryEnabled ? Date.now() : void 0;
				try {
					const result = await originalMethod.apply(instance, args);
					if (telemetryEnabled && startTime !== void 0) {
						const durationMs = Date.now() - startTime;
						const telemetryPayload = {
							functionName: `${instrumentedClassName}.${name}`,
							durationMs,
							timestamp: startTime / 1e3,
							props: {
								fileName: instrumentedClassName,
								method: name
							},
							metadata: { provider: this.telemetryMetadata?.provider ?? "openai" },
							error: void 0,
							source: this.telemetrySource
						};
						this.batchProcessor.pushItem(telemetryPayload);
					}
					return result;
				} catch (error) {
					if (error instanceof Error) {
						if (!error.errorId) {
							error.errorId = getRandomUUID();
							if (telemetryEnabled && startTime !== void 0) {
								const durationMs = Date.now() - startTime;
								this.prepareAndSendErrorTelemetry(error, instrumentedClassName, name, startTime, durationMs);
							}
						}
					}
					throw error;
				}
			};
		}
		return instance;
	}
	/**
	* Check if the telemetry should be sent.
	* @returns true if the telemetry should be sent, false otherwise
	*/
	shouldSendTelemetry() {
		const telemetryDisabledEnvironments = ["test", "ci"];
		const nodeEnv = (getEnvVariable("NODE_ENV", "development") || "").toLowerCase();
		const isDisabledEnvironment = telemetryDisabledEnvironments.includes(nodeEnv);
		const isTelemetryDisabledByEnv = getEnvVariable("TELEMETRY_DISABLED", "false") === "true";
		return !this.isTelemetryDisabled && !isTelemetryDisabledByEnv && !isDisabledEnvironment;
	}
	/**
	* Prepare and send the error telemetry.
	*
	* @param {unknown} error - The error to send.
	* @param {string} instrumentedClassName - The class name of the instrumented class.
	* @param {string} name - The name of the method that threw the error.
	* @param {number} startTime - The start time of the method invocation in milliseconds.
	* @param {number} durationMs - The duration of the method invocation in milliseconds.
	*/
	prepareAndSendErrorTelemetry(error, instrumentedClassName, name, startTime, durationMs) {
		const telemetryPayload = {
			functionName: `${instrumentedClassName}.${name}`,
			durationMs,
			timestamp: startTime / 1e3,
			props: {
				fileName: instrumentedClassName,
				method: name
			},
			metadata: { provider: this.telemetryMetadata?.provider ?? "openai" },
			source: this.telemetrySource
		};
		if (error instanceof ComposioError) telemetryPayload.error = {
			errorId: error.errorId,
			name: error.name,
			message: redactSensitiveText(error.message),
			stack: redactSensitiveText(error.stack)
		};
		else if (error instanceof ComposioError$1) telemetryPayload.error = {
			errorId: error.errorId,
			name: error.name,
			code: error.code,
			message: redactSensitiveText(error.message),
			stack: redactSensitiveText(error.stack)
		};
		else if (error instanceof Error) telemetryPayload.error = {
			errorId: error.errorId,
			name: error.name ?? "Unknown error",
			message: redactSensitiveText(error.message),
			stack: redactSensitiveText(error.stack)
		};
		this.deferTelemetry(() => this.sendErrorTelemetry(telemetryPayload));
	}
	deferTelemetry(callback) {
		const pending = new Promise((resolve) => {
			const run = () => {
				Promise.resolve().then(callback).catch((error) => {
					logger.debug("Error sending deferred telemetry", error);
				}).finally(resolve);
			};
			if (typeof queueMicrotask === "function") queueMicrotask(run);
			else setTimeout(run, 0);
		});
		this.pendingImmediateTelemetry.add(pending);
		pending.finally(() => {
			this.pendingImmediateTelemetry.delete(pending);
		});
	}
	async flushImmediateTelemetry() {
		while (this.pendingImmediateTelemetry.size > 0) await Promise.all(Array.from(this.pendingImmediateTelemetry));
	}
	/**
	* Send the telemetry payload to the server.
	* @param payload - the telemetry payload to send
	* @returns
	*/
	async sendMetric(payload) {
		if (!this.shouldSendTelemetry()) {
			logger.debug("Telemetry is disabled, skipping metric telemetry", payload);
			return;
		}
		try {
			logger.debug("SDK Metric", payload);
			await TelemetryService.sendMetric(payload);
		} catch (error) {
			logger.error("Error sending metric telemetry", error);
		}
	}
	async sendErrorTelemetry(payload) {
		if (!this.shouldSendTelemetry()) {
			logger.debug("Telemetry is disabled, skipping metric telemetry", payload);
			return;
		}
		try {
			logger.debug("SDK Error Telemetry", payload);
			await TelemetryService.sendErrorLog(payload);
		} catch (error) {
			logger.error("Error sending error telemetry", error);
		}
	}
	/**
	* Flush any pending telemetry and wait for it to complete.
	* This is automatically called on process exit in Node.js environments.
	*/
	async flush() {
		await this.batchProcessor.flush();
		await this.flushImmediateTelemetry();
	}
	/**
	* Register process exit handlers to automatically flush telemetry.
	* Only registers handlers in Node.js environments (not in browsers).
	*/
	registerExitHandlers() {
		if (this.exitHandlersRegistered || typeof process === "undefined" || !process.on) return;
		this.exitHandlersRegistered = true;
		const flushSync = () => {
			this.flush().catch((error) => {
				logger.debug("Error flushing telemetry on exit", error);
			});
		};
		process.on("beforeExit", () => {
			flushSync();
		});
		const createSignalHandler = (signal) => {
			const handler = () => {
				logger.debug(`Received ${signal}, flushing telemetry...`);
				this.flush().catch((error) => {
					logger.debug("Error flushing telemetry on signal", error);
				}).finally(() => {
					process.removeListener(signal, handler);
					process.kill(process.pid, signal);
				});
			};
			return handler;
		};
		process.on("SIGINT", createSignalHandler("SIGINT"));
		process.on("SIGTERM", createSignalHandler("SIGTERM"));
	}
};
const telemetry = new TelemetryTransport();

//#endregion
export { telemetry as t };
