// @composio/client/src/resources/tool-router/tool-router.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../core/resource';
import * as SessionAPI from './session/session';
import {
  Session,
  SessionAttachParams,
  SessionAttachResponse,
  SessionConfigHistoryParams,
  SessionConfigHistoryResponse,
  SessionCreateParams,
  SessionCreateResponse,
  SessionExecuteMetaParams,
  SessionExecuteMetaResponse,
  SessionExecuteParams,
  SessionExecuteResponse,
  SessionLinkParams,
  SessionLinkResponse,
  SessionPatchParams,
  SessionPatchResponse,
  SessionProxyExecuteParams,
  SessionProxyExecuteResponse,
  SessionRetrieveResponse,
  SessionSearchParams,
  SessionSearchResponse,
  SessionToolkitsParams,
  SessionToolkitsResponse,
  SessionToolsParams,
  SessionToolsResponse,
} from './session/session';

export class ToolRouter extends APIResource {
  session: SessionAPI.Session = new SessionAPI.Session(this._client);
}

ToolRouter.Session = Session;

export declare namespace ToolRouter {
  export {
    Session as Session,
    type SessionCreateResponse as SessionCreateResponse,
    type SessionRetrieveResponse as SessionRetrieveResponse,
    type SessionAttachResponse as SessionAttachResponse,
    type SessionConfigHistoryResponse as SessionConfigHistoryResponse,
    type SessionExecuteResponse as SessionExecuteResponse,
    type SessionExecuteMetaResponse as SessionExecuteMetaResponse,
    type SessionLinkResponse as SessionLinkResponse,
    type SessionPatchResponse as SessionPatchResponse,
    type SessionProxyExecuteResponse as SessionProxyExecuteResponse,
    type SessionSearchResponse as SessionSearchResponse,
    type SessionToolkitsResponse as SessionToolkitsResponse,
    type SessionToolsResponse as SessionToolsResponse,
    type SessionCreateParams as SessionCreateParams,
    type SessionAttachParams as SessionAttachParams,
    type SessionConfigHistoryParams as SessionConfigHistoryParams,
    type SessionExecuteParams as SessionExecuteParams,
    type SessionExecuteMetaParams as SessionExecuteMetaParams,
    type SessionLinkParams as SessionLinkParams,
    type SessionPatchParams as SessionPatchParams,
    type SessionProxyExecuteParams as SessionProxyExecuteParams,
    type SessionSearchParams as SessionSearchParams,
    type SessionToolkitsParams as SessionToolkitsParams,
    type SessionToolsParams as SessionToolsParams,
  };
}
